<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Momocs speed dating • Momocs</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Momocs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Momocs_grindr.html">Layered shape and multivariate plots with grindr</a>
    </li>
    <li>
      <a href="../articles/Momocs_speed_dating.html">Momocs speed dating</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MomX/Momocs">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Momocs speed dating</h1>
                        <h4 class="author">Vincent Bonhomme</h4>
            
            <h4 class="date">2018-03-12</h4>
          </div>

    
    
<div class="contents">
<div id="preliminaries" class="section level2">
<h2 class="hasAnchor">
<a href="#preliminaries" class="anchor"></a>Preliminaries</h2>
<div id="abstract" class="section level3">
<h3 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h3>
<p>Momocs aims to provide a complete and convenient toolkit for morphometrics. It is intended for scientists interested in describing quantitatively the shape, and its (co)variations, of the objects they study.</p>
<p>In the last decade, R has become the open-source <em>lingua franca</em> for statistics, and morphometrics known its so-called “revolution”. Nevertheless, morphometric analyses still have to be carried out using various software packages either dedicated to a particular morphometric and/or for which source code is mostly unavailable and/or copyrighted. Moreover, most of existing software packages cannot be extended and their bugs are hard to detect and thus correct. This situation is detrimental to morphometrics: time is wasted, analyses are restricted to available methods, and last but not least, are poorly reproducible. This impedes collaborative effort both in software development and in morphometric studies.</p>
<p>By gathering the common morphometric approaches in an open-source environment and welcoming contributions, Momocs is an attempt to solve this twofold problem.</p>
<p>Momocs hinges on the core functions published in the must-have <em><a href="http://www.springer.com/us/book/9780387777894">Morphometrics with R</a></em> by <a href="http://www.isem.univ-montp2.fr/recherche/equipes/macroevolution-et-developpement/personnel/claude-julien/">Julien Claude</a> (2008), but has been further extended to allow other shape description systems. So far, configurations of landmarks, outlines and open outline analyses, along with some facilities for traditional morphometrics are implemented.</p>
<p>Prior to analysis, Momocs can be used to acquire and manipulate data or to import/export from/to other formats. Momocs also has the facility for a wide range of multivariate analyses and production of the companion graphics. Thus a researcher will find that just a few lines of code will provide initial results, but the methods implemented can be finely tuned and extended according to the user’s needs.</p>
<ul>
<li>If you use it, cite it: <code>citation("Momocs")</code>.</li>
<li>This citation refers to an obsolete version of Momocs, only handling outline analyses. The next companion and seminal paper is submitted.</li>
</ul>
</div>
<div id="survival-tips" class="section level3">
<h3 class="hasAnchor">
<a href="#survival-tips" class="anchor"></a>Survival tips</h3>
<ul>
<li>This vignette introduces Momocs; more specific help can be find in function’s help files, for instance <code><a href="../reference/efourier.html">?efourier</a></code>.</li>
<li>There is a much nicer <a href="http://momx.github.io/Momocs/">online version of this manual</a> that can be accessed from the console with, e.g. <code><a href="../reference/Momocs_help.html">Momocs_help("efourier")</a></code>.</li>
<li>Feel free to contribute to Momocs through <a href="http://www.github.com/MomX/Momocs">GitHub</a>: report issues, ask for new features, share data and methods, correct typos, write better vignettes, helpfiles, or whatever pleases you. If you have never heard of GitHub, that’s definitely worth <a href="http://r-pkgs.had.co.nz/git.html">a look</a>.</li>
<li>Feel free to drop me a line, should you need a hand or would like to collaborate with me: <code>bonhomme.vincent@gmail.com</code>.</li>
</ul>
</div>
<div id="get-install-and-use-it" class="section level3">
<h3 class="hasAnchor">
<a href="#get-install-and-use-it" class="anchor"></a>Get, install and use it</h3>
<p>First, of all, let’s download the last version of Momocs. You will need to install the <code>devtools</code> package to get it from my <a href="http://www.github.com/MomX/Momocs">GitHub</a> repository :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"MomX/Momocs"</span>)</code></pre></div>
<p>The typical <code>install_packages("Momocs")</code> will get you the last CRAN version of Momocs, but the GitHub version is preferred as Momocs is still under active development.</p>
<p>We can start using Momocs, as long as it has been loaded using:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Momocs)</code></pre></div>
</div>
</div>
<div id="design" class="section level2">
<h2 class="hasAnchor">
<a href="#design" class="anchor"></a>Design</h2>
<p>Keywords used all accross Momocs are introduced here in <strong>bold</strong>.</p>
<p>Morphometrics is the ugly job of turning beautiful shapes into quantitative variables. Just kidding, that’s pretty exciting.</p>
<p>A <strong>shape</strong> is defined as a collection of <code>(x; y)</code> coordinates. No 3D yet but different <strong>families</strong> can be handled: <strong>outlines</strong>, here in a first-quarter moon ; <strong>open outlines</strong>, here is the sterile valve of an olive stone; <strong>configuration of landmarks</strong>; here, hologous points from a mosquito wing.</p>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-4-1.png" width="672"><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-4-2.png" width="672"><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-4-3.png" width="672"></p>
<p>They are all single shapes defined by a matrix of <code>(x; y)</code> coordinates; here are the first points of the moon:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shapes[<span class="dv">18</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()</code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]  200   50
## [2,]  199   49
## [3,]  198   49
## [4,]  197   50
## [5,]  196   50
## [6,]  195   49</code></pre>
<p>A few dozens of operations on single shapes are implemented such as: plotting, centering, calculating areas, etc. These 70+ operations can be accessed with <code>apropos("coo_")</code>. But working on single shapes is quite boring.</p>
<p>Shapes can be organized into a <strong>collection</strong> of coordinates: a <code>Coo</code> object that carries:</p>
<ul>
<li>a component named <code>$coo</code>, a <code>list</code> of shapes (as <code>matrix</code>.ces);</li>
<li>most of the time, a component named <code>$fac</code>, a <code>data.frame</code> to store covariates, either <code>factor</code>s or <code>numeric</code>s;</li>
<li>possibly, other components of interest.</li>
</ul>
<p>One can do several things with a <code>Coo</code> object: visualize it, apply morphometric operations, handle the data it contains, but in the end, a __ morphometric method__ will turn coordinates into coefficients.</p>
<p>Such morphometric operation on coordinates produce a collection of coefficients: a <code>Coe</code> object that carries:</p>
<ul>
<li>a component named <code>$coe</code>, a <code>matrix</code> of coefficients;</li>
<li>if present in <code>Coo</code>, <code>$fac</code> is inherited;</li>
<li>possibly, other components of interest.</li>
</ul>
<p>This can be summarized as follows:</p>
<table class="table">
<thead><tr class="header">
<th><code>Coo</code></th>
<th>+</th>
<th>Morphometric method</th>
<th>=</th>
<th><code>Coe</code></th>
</tr></thead>
<tbody><tr class="odd">
<td>
<code>(x; y)</code> coordinates</td>
<td>+</td>
<td>appropriate method</td>
<td>=</td>
<td>quantitative variables</td>
</tr></tbody>
</table>
<p><code>Coo</code> objects are collections of coordinates that become <code>Coe</code> objects when an appropriate morphometric method is applied on them.</p>
<p>Some operations on <code>Coo</code>/<code>Coe</code> are <strong>generic</strong> in that they do not depend of the nature of the shape. For instance, centering a configuration of landmarks or an outline, or calculating their centroid size is, mathematically, the same <strong>generic</strong> operation. But some operations on shapes are <strong>specific</strong> to a peculiar family. For instance, calculating elliptical Fourier transforms on a configuration of landmarks would make no sense.</p>
<p>Momocs implement this desirable behavior and defines <strong>classes</strong> and <strong>subclasses</strong>, as S3 objects.</p>
<table class="table">
<thead><tr class="header">
<th><code>Coo</code></th>
<th>Morphometrics methods</th>
<th><code>Coe</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>OutCoo</code> (outlines)</td>
<td>
<code>efourier</code>, <code>rfourier</code>, <code>sfourier</code>, <code>tfourier</code>
</td>
<td><code>OutCoe</code></td>
</tr>
<tr class="even">
<td>
<code>OpnCoo</code> (open outlines)</td>
<td>
<code>npoly</code>, <code>opoly</code>, <code>dfourier</code>
</td>
<td><code>OpnCoe</code></td>
</tr>
<tr class="odd">
<td>
<code>LdkCoo</code> (configuration of landmarks)</td>
<td>
<code>fgProcrustes</code>, <code>slide</code>
</td>
<td><code>LdkCoe</code></td>
</tr>
</tbody>
</table>
<p>In other words:</p>
<ul>
<li>any collection of shapes belongs to (pick one) <code>{OutCoo, OpnCoo, LdkCoo}</code> and is also a <code>Coo</code> object;</li>
<li>generic and specific methods can be applied to it</li>
<li>a collection of coefficients is obtain and belongs to (pick one) <code>{OutCoe, OpnCoe, LdkCoe}</code> and is also a <code>Coe</code> object.</li>
</ul>
<p>Finally, generic and specific operations can be applied to the <code>Coe</code> objects, chiefly multivariate methods, capitalicized: <code>PCA</code>, <code>LDA</code>, <code>CLUST</code>, <code>MANOVA</code>(and MANCOVA), <code>MSHAPES</code>, <code>KMEANS</code>, etc.</p>
<p>Overall, Momocs implements a simple and consistent grammar that is detailed below. Also, if you’re familiar with modern R and the Hadley-verse, you should feel home as <code>ggplot2</code> graphics, <code>dplyr</code> verbs and <code>magrittr</code> pipes are implemented.</p>
</div>
<div id="single-shapes" class="section level2">
<h2 class="hasAnchor">
<a href="#single-shapes" class="anchor"></a>Single shapes</h2>
<p>Let’s load one of the Momocs datasets, some various outlines (an <code>Out</code> object):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shapes                    <span class="co"># prints a brief summary</span>
<span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(shapes, <span class="dt">names=</span><span class="ot">TRUE</span>) <span class="co"># base graphics</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/panel2.html">panel2</a></span>(<span class="kw"><a href="../reference/slice.html">slice</a></span>(shapes, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)) <span class="co"># ggplot2 graphics on a subset (for speed sake)</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-6-2.png" width="672"></p>
<pre><code>## An Out object with: 
##  - $coo: 30 outlines (836 +/- 255 coordinates, all closed)
##  - $fac: No classifier defined in $fac</code></pre>
<p><code>shapes</code> is one of the datasets bundled with Momocs. It’s (“lazy”) loaded in memory as soon as you call it, no need for <code>data(shapes)</code>. To see all Momocs’ datasets, try <code>data(package="Momocs")</code>. These datasets are all <code>Coo</code> objects (try <code>class(bot)</code>), ie collection of shapes.</p>
<p>One can do many things on a <code>Coo</code> object, as above, e.g. printing a summary of it (just by typing its name in the console), plotting a family picture with <code>panel</code> or <code>panel2</code>. Note the <code>2</code> that refers to a <code>ggplot2</code> variant of a given plot.</p>
<p>So far, we’re interested in single shapes so let’s extract the 4th shape from <code>shapes</code>, using the traditional syntax. We plot it with <code>coo_plot</code> that comes with several options for plotting all families of shapes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shp &lt;-<span class="st"> </span>shapes[<span class="dv">4</span>]
<span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(shp)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># coo_plot is the base plotter for shapes</span>
<span class="co"># but it can be finely customized, see ?coo_plot</span>
<span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(shp, <span class="dt">col=</span><span class="st">"grey80"</span>, <span class="dt">border=</span><span class="ot">NA</span>, <span class="dt">centroid=</span><span class="ot">FALSE</span>, <span class="dt">main=</span><span class="st">"Meow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-7-2.png" width="672"></p>
<p>Let’s now do some basic operations on this shape. They all named <code>coo_*</code> and you can have the full list with <code>apropos("coo_")</code>. <code>coo_*</code> family encompasses:</p>
<ul>
<li>geometric operations (such as centering, scaling, etc.)</li>
<li>plotting functions</li>
<li>scalar descriptors of shape (such as area, perimeter, circularity, rectilinearity, etc.)</li>
<li>various other operations on a single shape.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(<span class="kw"><a href="../reference/coo_center.html">coo_center</a></span>(shp), <span class="dt">main=</span><span class="st">"centered Meow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(<span class="kw"><a href="../reference/coo_sample.html">coo_sample</a></span>(shp, <span class="dv">64</span>), <span class="dt">points=</span><span class="ot">TRUE</span>, <span class="dt">pch=</span><span class="dv">20</span>, <span class="dt">main=</span><span class="st">"64-pts Meow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-8-2.png" width="672"></p>
<p>Momocs makes use of <a href="https://CRAN.R-project.org/package=magrittr">maggritr</a>’s pipe operators. A nice introduction can be found <a href="https://CRAN.R-project.org/package=magrittr">there</a>. <code>magrittr</code> requires a (very small) cerebral gymnastics at the beginning but the benefits are huge, for defining moprhometric pipelines in Momocs but also for R as a whole. It makes things clearer, it: saves typing; reduces intermediate variable assignation; reads from left to right; substantiates the pipe we (should) have in mind. <code>magrittr</code>’s pipes are already loaded with Momocs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shapes[<span class="dv">4</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_smooth.html">coo_smooth</a></span>(<span class="dv">5</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_sample.html">coo_sample</a></span>(<span class="dv">64</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">coo_scale</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># pipes can be turned into custom function</span>
cs64 &lt;-<span class="st"> </span><span class="cf">function</span>(x) x <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_sample.html">coo_sample</a></span>(<span class="dv">64</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">coo_scale</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_center.html">coo_center</a></span>()
shapes[<span class="dv">4</span>] <span class="op">%&gt;%</span><span class="st"> </span>cs64 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>() <span class="co"># note the axes</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-9-2.png" width="672"></p>
<p>The most familiar operation can directly be applied on <code>Coo</code> objects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>coo_center <span class="op">%&gt;%</span><span class="st"> </span>coo_scale <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/coo_alignxax.html">coo_alignxax</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_slidedirection.html">coo_slidedirection</a></span>(<span class="st">"up"</span>) <span class="op">%T&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">print</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">stack</span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<pre><code>## An Out object with: 
##  - $coo: 40 outlines (162 +/- 21 coordinates, all unclosed)
##  - $fac: 1 classifier:
##      'type' (factor 2): beer, whisky.</code></pre>
</div>
<div id="morphometrics" class="section level2">
<h2 class="hasAnchor">
<a href="#morphometrics" class="anchor"></a>Morphometrics</h2>
<div id="outline-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#outline-analysis" class="anchor"></a>Outline analysis</h3>
<p>A word about data import: you can extract outlines from a list of black masks over a white background, as <code>.jpg</code> images with <code>import_jpg</code>. Have a look to helpfiles (<code>import_jpg</code> and <code>import_jpg1</code>) for more details. Here we do not bother with import since we will use the <code>bot</code>tles outlines dataset bundled with Momocs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(bot)
bot
<span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(bot, <span class="dt">fac=</span><span class="st">"type"</span>, <span class="dt">names=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">stack</span>(bot)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-11-2.png" width="672"></p>
<pre><code>## An Out object with: 
##  - $coo: 40 outlines (162 +/- 21 coordinates, all unclosed)
##  - $fac: 1 classifier:
##      'type' (factor 2): beer, whisky.</code></pre>
<p>Here we will illustrate outline analysis with some elliptical Fourier transforms (but the less used - and tested - radii variation Fourier transforms, its variant used by Renaud et al., and tangent angle Fourier transforms are also implemented with <code>rfourier</code>, <code>sfourier</code> and <code>tfourier</code> respectively).</p>
<p>The idea behind elliptical Fourier transforms is to fit the x and y coordinates separately, that is the blue and red curves below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_oscillo.html">coo_oscillo</a></span>(bot[<span class="dv">1</span>], <span class="st">"efourier"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<pre><code>##      dx   dy
## 1     0    0
## 2     3  -21
## 3     3  -32
## 4     6  -53
## 5     9  -74
## 6    11  -84
## 7    15 -105
## 8    17 -126
## 9    20 -147
## 10   22 -158
## 11   24 -179
## 12   26 -200
## 13   27 -211
## 14   27 -232
## 15   27 -253
## 16   27 -263
## 17   25 -284
## 18   23 -305
## 19   21 -316
## 20   17 -337
## 21   13 -358
## 22   11 -368
## 23    8 -389
## 24    5 -411
## 25    3 -432
## 26    3 -442
## 27    3 -463
## 28    7 -484
## 29   10 -495
## 30   22 -516
## 31   43 -532
## 32   53 -536
## 33   74 -542
## 34   95 -544
## 35  106 -546
## 36  127 -547
## 37  148 -546
## 38  159 -546
## 39  180 -545
## 40  201 -543
## 41  222 -537
## 42  232 -534
## 43  253 -521
## 44  270 -501
## 45  274 -490
## 46  278 -469
## 47  279 -448
## 48  279 -437
## 49  278 -416
## 50  278 -395
## 51  278 -385
## 52  276 -364
## 53  275 -343
## 54  274 -332
## 55  274 -311
## 56  271 -290
## 57  269 -269
## 58  268 -259
## 59  266 -237
## 60  263 -216
## 61  262 -206
## 62  260 -185
## 63  257 -164
## 64  255 -153
## 65  254 -132
## 66  253 -111
## 67  253 -101
## 68  254  -80
## 69  255  -59
## 70  258  -38
## 71  260  -27
## 72  264   -6
## 73  268   15
## 74  270   26
## 75  273   47
## 76  274   68
## 77  273   78
## 78  273   99
## 79  270  120
## 80  267  131
## 81  259  152
## 82  247  172
## 83  241  183
## 84  231  204
## 85  223  225
## 86  217  246
## 87  214  256
## 88  211  277
## 89  209  298
## 90  208  309
## 91  205  330
## 92  203  351
## 93  202  361
## 94  200  383
## 95  198  404
## 96  197  414
## 97  194  435
## 98  192  456
## 99  191  467
## 100 196  488
## 101 192  507
## 102 195  526
## 103 189  535
## 104 168  540
## 105 146  541
## 106 136  541
## 107 115  540
## 108  94  539
## 109  83  537
## 110  68  521
## 111  75  504
## 112  72  494
## 113  74  473
## 114  73  452
## 115  73  441
## 116  71  420
## 117  68  399
## 118  66  378
## 119  66  367
## 120  64  346
## 121  61  325
## 122  59  315
## 123  59  294
## 124  57  273
## 125  56  262
## 126  52  241
## 127  45  220
## 128  42  210
## 129  32  188
## 130  22  167
## 131  17  157
## 132   8  137
## 133   1  116
## 134  -3   95
## 135  -4   84
## 136  -4   63
## 137  -3   42
## 138  -2   32</code></pre>
<p>Graphically, this is equivalent to fitting Ptolemaic ellipses on the plane:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/Ptolemy.html">Ptolemy</a></span>(bot[<span class="dv">1</span>])</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>Let’s calibrate the number of harmonics required. More details can be found in their respective help files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/calibrate_harmonicpower.html">calibrate_harmonicpower</a></span>(bot)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/calibrate_deviations.html">calibrate_deviations</a></span>(bot)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-14-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/calibrate_reconstructions.html">calibrate_reconstructions</a></span>(bot)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-14-3.png" width="672"></p>
<pre><code>##   90%   95%   99% 99.9% 
##     5     5    10    17</code></pre>
<p>Here, 10 harmonics gather 99% of the harmonic power. If you’re happy with this criterium, you can even omit <code>nb.h</code> in <code>efourier</code>: that’s the default parameter, returned with a message.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.f &lt;-<span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(bot, <span class="dt">nb.h=</span><span class="dv">10</span>)
bot.f</code></pre></div>
<pre><code>## An OutCoe object [ elliptical Fourier analysis ]
## --------------------
##  - $coe: 40 outlines described, 10 harmonics
##  - $fac: 1 classifier:
##      'type' (factor 2): beer, whisky.</code></pre>
<p><code>bot.f</code> is a <code>Coe</code> object (and even an <code>OutCoe</code>), you have have a look to the help files to go deeper into Momocs classes.</p>
<p>Let’s have a look to the amplitude of fitted coefficients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(bot.f, <span class="dt">drop=</span><span class="dv">0</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">boxplot</span>(bot.f, <span class="dt">drop=</span><span class="dv">1</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-16-2.png" width="672"></p>
<p>Now, we can calculate a PCA on the <code>Coe</code> object and plot it, along with morphospaces, calculated on the fly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>(bot.f)
<span class="kw">class</span>(bot.p)        <span class="co"># a PCA object, let's plot it</span>
<span class="kw">plot</span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<pre><code>## [1] "PCA"    "prcomp"</code></pre>
<p>Amazing but we will do much better afterwards.</p>
<p>The question of normalization in elliptical Fourier transforms is central. Normalization can either be done beforehand, with geometric operations, or afterhand, directly on the matrix of Fourier coefficients and consuming the first harmonic. In brief, I’m not a big fan of the “use the first harmonics and see what happens” strategy, as some biases can be introduced (and actually quite hard to detect), particularly on rounded/ellipsoid shapes. More can be found in <code><a href="../reference/efourier.html">?efourier</a></code>.</p>
<p>Here is an example of such bias from the <code>molars</code> dataset generously shared by Cornu and Detroit.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># raw molars dataset</span>
<span class="kw">stack</span>(molars, <span class="dt">title =</span> <span class="st">"Non-aligned molars"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Procrustes-aligned and slided molars</span>
mol.al &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fgProcrustes.html">fgProcrustes</a></span>(molars, <span class="dt">tol =</span> <span class="fl">1e-4</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_slidedirection.html">coo_slidedirection</a></span>(<span class="st">"left"</span>)
<span class="kw">stack</span>(mol.al, <span class="dt">title=</span><span class="st">"Aligned molars"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Now compare PCA and morphospace using the 1st harmonic alignment</span>
molars <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dt">norm=</span><span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"type"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-3.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># and the a priori normalization </span>
molars <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dt">norm=</span><span class="ot">FALSE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"type"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-4.png" width="672"></p>
<p>Finally, you can drop some harmonics with <code>rm_harm</code>. And methods that removes the bilateral (a)symmetry are implemented: <code>rm_asym</code> and <code>rm_sym</code>, while <code>symmetry</code> calculates some related indices.</p>
</div>
<div id="open-outlines" class="section level3">
<h3 class="hasAnchor">
<a href="#open-outlines" class="anchor"></a>Open outlines</h3>
<p>Open outlines are curves. Methods actually implemented are:</p>
<ul>
<li>
<code>npoly</code> that fit natural polynomials;</li>
<li>
<code>opoly</code> that fit orthogonal (also called Legendre’s) polynomials;</li>
<li>
<code>dfourier</code> for the discrete cosine transform.</li>
</ul>
<p>Note that <code>opoly</code> and <code>npoly</code> can only be used on <em>simple</em> curves, curves that have at most one <code>y</code> for any <code>x</code> coordinates, at least under a given orientation. <code>dfourier</code> can fit <em>complex</em> curves, curves “that back on their feets”.</p>
<p>Here, we will work on the fertile valves of olive stones, a (very partial) dataset provided by my colleagues Terral, Ivorra, and others.</p>
<p>They have two orthogonal views (a lateral and a dorsal view). See the paper cited in <code><a href="../reference/data_olea.html">?olea</a></code> for more details. Let’s explore it a bit:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olea
<span class="kw">stack</span>(olea, <span class="dt">fac=</span><span class="st">"view"</span>)     <span class="co"># already aligned \o/</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(olea, <span class="dt">names=</span><span class="st">"ind"</span>)    <span class="co"># another family picture</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-19-2.png" width="672"></p>
<pre><code>## An Opn object with: 
##  - $coo: 210 open outlines (99 +/- 4 coordinates)
##  - $fac: 4 classifiers:
##      'var' (factor 4): Aglan, Cypre, MouBo1, PicMa.
##      'domes' (factor 2): cult, wild.
##      'view' (factor 2): VD, VL.
##      'ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.</code></pre>
<p>Now, we gonna calculate <code>opoly</code> on it and plot the result of the PCA. Notice how consistent is the grammar and the objects obtained:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">op &lt;-<span class="st"> </span><span class="kw"><a href="../reference/opoly.html">opoly</a></span>(olea)           <span class="co"># orthogonal polynomials</span>
<span class="kw">class</span>(op)                   <span class="co"># an OpnCoe, but also a Coe</span>
op.p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>(op)             <span class="co"># we calculate a PCA on it</span>
<span class="kw">class</span>(op.p)                 <span class="co"># a PCA object</span>
<span class="kw">plot</span>(<span class="kw"><a href="../reference/PCA.html">PCA</a></span>(op), <span class="op">~</span>domes<span class="op">+</span>var)   <span class="co"># notice the formula interface to combine factors</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<pre><code>## [1] "OpnCoe" "Coe"   
## [1] "PCA"    "prcomp"</code></pre>
<p>But this is perfectly wrong! We merged the two views are if they were different individuals. Momocs can first <code>chop</code> or <code>filter</code> the whole dataset to separate the two views, do morphometrics on them, and <code>combine</code> them afterwards.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/table.html">table</a></span>(olea, <span class="st">"view"</span>, <span class="st">"var"</span>) 
<span class="co"># we drop 'Cypre' since there is no VL for 'Cypre' var</span>
olea <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filter.html">filter</a></span>(var <span class="op">!=</span><span class="st"> "Cypre"</span>) <span class="op">%&gt;%</span><span class="st">              </span>
<span class="st">  </span><span class="co"># split, do morphometrics, combine</span>
<span class="st">  </span><span class="kw"><a href="../reference/chop.html">chop</a></span>(view) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">lapply</span>(opoly) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/combine.html">combine</a></span>() <span class="op">%T&gt;%</span>
<span class="st">   </span><span class="co"># we print the OpnCoe object, then resume to the pipe</span>
<span class="st">  </span><span class="kw">print</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># note the two views in the morphospace</span>
<span class="st">  </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"var"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<pre><code>##     var
## view Aglan Cypre MouBo1 PicMa
##   VD    30    30     30    30
##   VL    30     0     30    30
## An OpnCoe object [ combined: opoly + opoly analyses ]
## --------------------
##  - $coe: 90 open outlines described
##  - $baseline1: (-0.5; 0; -0.5; 0)
##  - $baseline2: (0.5; 0; 0.5; 0)
##  - $fac: 4 classifiers:
##      'var' (factor 3): Aglan, MouBo1, PicMa.
##      'domes' (factor 2): cult, wild.
##      'view' (factor 2): VD, VL.
##      'ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.</code></pre>
<p>Now the PCA is done on the combination of two <code>OpnCoe</code> objects, each one resulting from an independant <code>opoly</code> call. That is the meaning of the <code>[ combined: opoly + opoly analyses ]</code> printed by the pipe above. Momocs can combine up to four different views.</p>
</div>
<div id="configuration-of-landmarks" class="section level3">
<h3 class="hasAnchor">
<a href="#configuration-of-landmarks" class="anchor"></a>Configuration of landmarks</h3>
<p>_Landmarks methods are still quite experimental (i.e. not tested extensively)</p>
<p>Let’s have a look to graphics facilities and apply a full generalized Procrustes adjustment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">stack</span>(wings, <span class="dt">title =</span> <span class="st">"Raw wings"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">w.al &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fgProcrustes.html">fgProcrustes</a></span>(wings)
<span class="kw">stack</span>(w.al, <span class="dt">title =</span> <span class="st">"Aligned wings"</span>)
<span class="kw"><a href="../reference/ldk_chull.html">ldk_chull</a></span>(w.al<span class="op">$</span>coo)
<span class="kw"><a href="../reference/ldk_labels.html">ldk_labels</a></span>(<span class="kw"><a href="../reference/mshapes.html">mshapes</a></span>(w.al<span class="op">$</span>coo))</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-22-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># try those as well</span>
<span class="co">#ldk_confell(w.al$coo, col = "red")</span>
<span class="co">#ldk_contour(w.al$coo)</span>

<span class="co"># PCA</span>
<span class="kw"><a href="../reference/PCA.html">PCA</a></span>(w.al) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="dv">1</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-22-3.png" width="672"></p>
<p>Sliding landmarks are supported and rely on <code>geomorph</code> package by D.C. Adams and E. Otarola-Castillo.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">stack</span>(chaff, <span class="dt">title=</span><span class="st">"Raw chaff"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-23-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chaff.al &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fgsProcrustes.html">fgsProcrustes</a></span>(chaff)
<span class="kw">stack</span>(chaff.al, <span class="dt">title=</span><span class="st">"Aligned chaff"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-23-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chaff.al <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="op">~</span>taxa, <span class="dt">chull.filled=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-23-3.png" width="672"></p>
<pre><code>## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |=============                                                    |  20%
  |                                                                       
  |==========================                                       |  40%
  |                                                                       
  |=======================================                          |  60%
  |                                                                       
  |====================================================             |  80%
  |                                                                       
  |=================================================================| 100%</code></pre>
<p>Again, the grammar is consistent for landmarks.</p>
</div>
<div id="traditional-morphometrics" class="section level3">
<h3 class="hasAnchor">
<a href="#traditional-morphometrics" class="anchor"></a>Traditional morphometrics</h3>
<p>Traditional morphometrics lose geometries: from the variables, you can’t unambiguously reconstruct the shape. Every shape is described by a combination of measurements, (inter landmark distance, quantitative variables, scalar descriptor, etc.)</p>
<p>Momocs provides some basics utilities to work with such objects in the <code>TraCoe</code> class. There is not <code>TraCoo</code> per se but it can be obtained from any <code>Coo</code> with the <code>measure</code> method. Let’s take the <code>hearts</code> dataset that comes from handdrawn heart shapes from my former colleagues at the French Intitute of Pondicherry:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hearts
<span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(hearts, <span class="dt">fac=</span><span class="st">"aut"</span>, <span class="dt">names=</span><span class="st">"aut"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-24-1.png" width="672"></p>
<pre><code>## An Out object with: 
##  - $coo: 240 outlines (80 +/- 0 coordinates, all unclosed)
##  - $ldk: 4 landmark(s) defined
##  - $fac: 1 classifier:
##      'aut' (factor 8): ced, jeya, mat, ponnu, remi, rom, ruks, vince.</code></pre>
<p>Notice that there are 4 landmarks defined on them. Such landmarks on outlines can be: defined with<code><a href="../reference/def_ldk.html">def_ldk()</a></code>, retrieved with <code><a href="../reference/get_ldk.html">get_ldk()</a></code>, and overall used to align outlines with <code><a href="../reference/fgProcrustes.html">fgProcrustes()</a></code>. You can compare: <code>hearts %&gt;% stack()</code> with <code>hearts %&gt;% fgProcrustes() %&gt;% coo_slide(ldk=1) %&gt;% stack()</code>.</p>
<p>Let’s describe these hearts with scalar descriptors: area, circularity and the distance between the 1st and the 3rd bumps of the hearts. <code>measure</code> is of great help. Note the loadings.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht &lt;-<span class="st"> </span><span class="kw"><a href="../reference/measure.html">measure</a></span>(hearts, coo_area, coo_circularity, <span class="kw"><a href="../reference/d.html">d</a></span>(<span class="dv">1</span>, <span class="dv">3</span>))
ht <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"aut"</span>, <span class="dt">pch=</span><span class="dv">20</span>, <span class="dt">ellipsesax=</span>F, <span class="dt">ellipse=</span>T, <span class="dt">loadings=</span>T)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-25-1.png" width="672"></p>
<p>Again, there are plenty of scalar descriptors of shape, which names starts with <code>coo_*</code>, <code>apropos("coo_")</code>.</p>
<p>Such a <code>TraCoe</code> is provided in the <code>flower</code> dataset which is simply a rearranged <code>iris</code>. Once again, note the grammar consistency.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">flower
flower <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"sp"</span>, <span class="dt">loadings=</span><span class="ot">TRUE</span>, <span class="dt">contour=</span><span class="ot">TRUE</span>, <span class="dt">lev.contour=</span><span class="dv">5</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-26-1.png" width="672"></p>
<pre><code>## A TraCoe object --------------------
##  - $coe: 150 shapes described with 4 variables
##  - $fac: 1 classifier:
##      'sp' (factor 3): setosa, versicolor, virginica.</code></pre>
<p>Note that, by default, PCA on <code>TraCoe</code> object first centers and scales variables. This can be changed, see <code><a href="../reference/PCA.html">?PCA</a></code>.</p>
</div>
</div>
<div id="multivariate-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#multivariate-statistics" class="anchor"></a>Multivariate statistics</h2>
<p>This section will mainly be illustrated with <code>bot</code>, and consequently outline analysis, but it works the same on any <code>Coe</code> object, resulting from open outlines, configuration of landmarks, traditional morphometrics, etc.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.f &lt;-<span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(bot)</code></pre></div>
<div id="pca-principal-component-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#pca-principal-component-analysis" class="anchor"></a><code>PCA</code>: Principal Component Analysis</h3>
<p>Let’s see the main components of shape variability with a Principal Component Analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>(bot.f)
<span class="kw">plot</span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-28-1.png" width="672"></p>
<p>Morphological spaces are reconstructed on the fly with <code>plot.PCA</code>. We call it <code>plot.PCA</code> because it uses the familiar <code>plot</code> but on the particular <code>PCA</code> class (type <code>class(bot.p)</code>). We may want to display the two groups saved in <code>bot$fac</code>. Just type the id of the column or its name.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="st">"type"</span>) <span class="co"># equivalent to plot(bot.p, 1)</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-29-1.png" width="672"></p>
<p><code>plot.PCA</code> has many arguments and examples below will give a glimpse, but see <code><a href="../reference/plot.PCA.html">?plot.PCA</a></code> for the full list.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="dv">1</span>, <span class="dt">ellipses=</span><span class="ot">TRUE</span>, <span class="dt">ellipsesax =</span> <span class="ot">FALSE</span>, <span class="dt">pch=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">5</span>))</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-30-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="dv">1</span>, <span class="dt">chull=</span><span class="ot">TRUE</span>, <span class="dt">pos.shp =</span> <span class="st">"full_axes"</span>, <span class="dt">abbreviate.labelsgroups =</span> <span class="ot">TRUE</span>, <span class="dt">points=</span><span class="ot">FALSE</span>, <span class="dt">labelspoints =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-30-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="dv">1</span>, <span class="dt">pos.shp=</span><span class="st">"circle"</span>, <span class="dt">stars=</span><span class="ot">TRUE</span>, <span class="dt">chull.filled=</span><span class="ot">TRUE</span>, <span class="dt">palette=</span>col_spring)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-30-3.png" width="672"></p>
<p>You have a <code>ggplot2</code> alternative, that is not completely consistent:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot2(bot.p, "type") # deprecated for the moment</span></code></pre></div>
<p>And other helper functions for the <code>PCA</code> class:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/scree.html">scree</a></span>(bot.p)
<span class="kw"><a href="../reference/scree.html">scree_plot</a></span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-32-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">boxplot</span>(bot.p, <span class="dv">1</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-32-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/PCcontrib.html">PCcontrib</a></span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-32-3.png" width="672"></p>
<pre><code>## # A tibble: 10 x 3
##    axis  proportion cumsum
##    &lt;ord&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1 1        0.761    0.761
##  2 2        0.170    0.931
##  3 3        0.0294   0.960
##  4 4        0.0135   0.974
##  5 5        0.00860  0.982
##  6 6        0.00719  0.990
##  7 7        0.00306  0.993
##  8 8        0.00190  0.994
##  9 9        0.00159  0.996
## 10 10       0.00122  0.997</code></pre>
<p>By the way, you can use Momocs plotters to plot non-morphometric datasets. Using a <code>TraCoe</code> object is an option, but <code>PCA</code> also works fine. Let’s see an example with <code>iris</code> dataset:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/TraCoe.html">TraCoe</a></span>(iris[, <span class="op">-</span><span class="dv">5</span>], <span class="dt">fac=</span><span class="kw">data.frame</span>(<span class="dt">sp=</span>iris<span class="op">$</span>Species)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"sp"</span>, <span class="dt">loadings=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-33-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># or</span>
<span class="kw"><a href="../reference/PCA.html">PCA</a></span>(iris[, <span class="op">-</span><span class="dv">5</span>], <span class="dt">fac=</span><span class="kw">data.frame</span>(<span class="dt">sp=</span>iris<span class="op">$</span>Species)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">plot</span>(<span class="st">"sp"</span>, <span class="dt">chull=</span><span class="ot">TRUE</span>, <span class="dt">ellipses=</span><span class="ot">TRUE</span>, <span class="dt">conf_ellipses =</span> <span class="fl">0.9</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-33-2.png" width="672"></p>
</div>
<div id="lda-linear-discriminant-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#lda-linear-discriminant-analysis" class="anchor"></a><code>LDA</code>: Linear Discriminant Analysis</h3>
<p>We can also calculate a Linear Discriminant Analysis on the PCA scores, or on the <code>Coe</code> object, directly on the matrix of coefficients (and results may be better yet we may encounter collinearity between variables). Try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#LDA(bot.f, 1)</span>
<span class="co"># we work on PCA scores</span>
bot.l &lt;-<span class="st"> </span><span class="kw"><a href="../reference/LDA.html">LDA</a></span>(bot.p, <span class="dv">1</span>)
<span class="co"># print a summary, along with the leave-one-out cross-validation table. </span>
bot.l 
<span class="co"># a much more detailed summary</span>
bot.l <span class="op">%&gt;%</span><span class="st"> </span>summary
<span class="co"># plot.LDA works pretty much with the same grammar as plot.PCA</span>
<span class="co"># here we only have one LD</span>
<span class="kw">plot</span>(bot.l)
<span class="co"># plot the cross-validation table</span>
<span class="kw"><a href="../reference/plot_CV.html">plot_CV</a></span>(bot.l)  <span class="co"># tabular version</span>
<span class="kw"><a href="../reference/plot_CV.html">plot_CV</a></span>(bot.l, <span class="dt">freq=</span><span class="ot">TRUE</span>) <span class="co"># frequency table</span>
<span class="kw"><a href="../reference/plot_CV2.html">plot_CV2</a></span>(bot.l) <span class="co"># arrays version</span></code></pre></div>
<p>Finally, if you need to classify using LDA methods, see <code>classify</code>.</p>
</div>
<div id="manova-multivariate-analysis-of-covariace" class="section level3">
<h3 class="hasAnchor">
<a href="#manova-multivariate-analysis-of-covariace" class="anchor"></a><code>MANOVA</code>: Multivariate Analysis of (co)variace</h3>
<p>We can test for a difference in the distribution of PC scores with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/MANOVA.html">MANOVA</a></span>(bot.p, <span class="st">"type"</span>)</code></pre></div>
<pre><code>##           Df Hotelling-Lawley approx F num Df den Df   Pr(&gt;F)    
## fac        1           2.8388   12.977      7     32 8.88e-08 ***
## Residuals 38                                                     
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
<p>We can also calculate pairwise combination between every levels of a <code>fac</code>. Here we just have two levels, so a single pairwise combination but the syntax is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/MANOVA_PW.html">MANOVA_PW</a></span>(bot.p, <span class="st">"type"</span>)</code></pre></div>
<pre><code>## $stars.tab
##      beer whisky
## beer      ***   
## 
## $summary (see also $manovas)
##               Df Pillai approx F num Df den Df   Pr(&gt;F)
## beer - whisky  1 0.7395    12.98      7     32 8.88e-08</code></pre>
<p>If we want a MANCOVA instead :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mutate.html">mutate</a></span>(<span class="dt">cs=</span><span class="kw"><a href="../reference/coo_centsize.html">coo_centsize</a></span>(.))
bot <span class="op">%&gt;%</span><span class="st"> </span>efourier <span class="op">%&gt;%</span><span class="st"> </span>PCA <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/MANOVA.html">MANOVA</a></span>(<span class="st">"cs"</span>)</code></pre></div>
<pre><code>##           Df Hotelling-Lawley approx F num Df den Df Pr(&gt;F)
## fac        1          0.32229   1.4733      7     32  0.212
## Residuals 38</code></pre>
</div>
<div id="clust-hierarchical-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#clust-hierarchical-clustering" class="anchor"></a><code>CLUST</code>: Hierarchical clustering</h3>
<p>A hierarchical classification now. It relies on <code>dist</code> + <code>hclust</code> + <code><a href="http://www.rdocumentation.org/packages/ape/topics/plot.phylo">ape::plot.phylo</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/CLUST.html">CLUST</a></span>(bot.p, <span class="dv">1</span>)</code></pre></div>
<p>Monophyly is plotted by default. Many options can be found in <code><a href="../reference/CLUST.html">?CLUST</a></code></p>
</div>
<div id="kmeans-k-means-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#kmeans-k-means-clustering" class="anchor"></a><code>KMEANS</code>: K-means clustering</h3>
<p>A very minimal k-means clustering is implemented:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/KMEANS.html">KMEANS</a></span>(bot.p, <span class="dt">centers =</span> <span class="dv">5</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-39-1.png" width="672"></p>
<pre><code>## K-means clustering with 5 clusters of sizes 6, 5, 5, 18, 6
## 
## Cluster means:
##           PC1          PC2
## 1  0.10184224 -0.028875902
## 2  0.02497106 -0.006671549
## 3  0.06730453  0.050371247
## 4 -0.04433327  0.008000365
## 5 -0.04573875 -0.031541607
## 
## Clustering vector:
##         brahma          caney         chimay         corona    deusventrue 
##              5              4              1              4              4 
##          duvel   franziskaner     grimbergen        guiness     hoegardeen 
##              1              5              2              2              4 
##        jupiler     kingfisher       latrappe lindemanskriek    nicechouffe 
##              4              4              1              4              4 
##     pecheresse   sierranevada     tanglefoot          tauro      westmalle 
##              4              2              1              4              4 
##          amrut    ballantines      bushmills         chivas        dalmore 
##              4              1              5              3              3 
##   famousgrouse    glendronach   glenmorangie   highlandpark    jackdaniels 
##              5              4              4              1              2 
##             jb  johnniewalker       magallan     makersmark           oban 
##              4              5              5              3              4 
##     oldpotrero      redbreast         tamdhu     wildturkey         yoichi 
##              3              3              4              4              2 
## 
## Within cluster sum of squares by cluster:
## [1] 0.006121728 0.001916822 0.002720829 0.006537571 0.001860776
##  (between_SS / total_SS =  89.4 %)
## 
## Available components:
## 
## [1] "cluster"      "centers"      "totss"        "withinss"    
## [5] "tot.withinss" "betweenss"    "size"         "iter"        
## [9] "ifault"</code></pre>
</div>
<div id="mshapes-mean-shapes" class="section level3">
<h3 class="hasAnchor">
<a href="#mshapes-mean-shapes" class="anchor"></a><code>mshapes</code>: Mean shapes</h3>
<p>We can retrieve the mean shapes, group wise (if a <code>fac</code> is specified), or the global mean shape (if omitted). It works from the <code>Coe</code> object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mean shape</span>
bot.f <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mshapes.html">mshapes</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-40-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mean shape, per group</span>
bot.ms &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mshapes.html">mshapes</a></span>(bot.f, <span class="dv">1</span>)
beer   &lt;-<span class="st"> </span>bot.ms<span class="op">$</span>shp<span class="op">$</span>beer    <span class="op">%T&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(<span class="dt">border=</span><span class="st">"blue"</span>)
whisky &lt;-<span class="st"> </span>bot.ms<span class="op">$</span>shp<span class="op">$</span>whisky  <span class="op">%T&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_draw.html">coo_draw</a></span>(<span class="dt">border=</span><span class="st">"red"</span>)
<span class="kw">legend</span>(<span class="st">"topright"</span>, <span class="dt">lwd=</span><span class="dv">1</span>,
       <span class="dt">col=</span><span class="kw">c</span>(<span class="st">"blue"</span>, <span class="st">"red"</span>), <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">"beer"</span>, <span class="st">"whisky"</span>))</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-40-2.png" width="672"></p>
<p>We can also plot a pairwise comparison of them:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">leaves &lt;-<span class="st"> </span>shapes <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/slice.html">slice</a></span>(<span class="kw">grep</span>(<span class="st">"leaf"</span>, <span class="kw">names</span>(shapes))) <span class="op">%$%</span><span class="st"> </span>coo
leaves <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/plot_mshapes.html">plot_mshapes</a></span>(<span class="dt">col2=</span><span class="st">"#0000FF"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-41-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># or from mshapes directly</span>
bot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dv">6</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mshapes.html">mshapes</a></span>(<span class="st">"type"</span>) <span class="op">%&gt;%</span><span class="st"> </span>plot_mshapes</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-41-2.png" width="672"></p>
</div>
</div>
<div id="manipulating-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#manipulating-objects" class="anchor"></a>Manipulating objects</h2>
<div id="dplyr-verbs" class="section level3">
<h3 class="hasAnchor">
<a href="#dplyr-verbs" class="anchor"></a><code>dplyr</code> verbs</h3>
<p>One common yet boring task of morphometrics consists in handling datasets: add new information, remove some individuals, etc.</p>
<p>Momocs adapts <a href="https://CRAN.R-project.org/package=dplyr">dplyr</a> verbs to its objects, and add new ones. If you have never heard of dplyr, let’s have a look to its introduction <a href="https://CRAN.R-project.org/package=dplyr">there</a>, this may change your (R) life.</p>
<p>Basics verbs are implemented; try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(olea)
olea</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 210 open outlines (99 +/- 4 coordinates)
##  - $fac: 4 classifiers:
##      'var' (factor 4): Aglan, Cypre, MouBo1, PicMa.
##      'domes' (factor 2): cult, wild.
##      'view' (factor 2): VD, VL.
##      'ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># slice: select individuals based on their position</span>
<span class="kw"><a href="../reference/slice.html">slice</a></span>(olea, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 5 open outlines (99 +/- 4 coordinates)
##  - $fac: 4 classifiers:
##      'var' (factor 1): Aglan.
##      'domes' (factor 1): cult.
##      'view' (factor 2): VD, VL.
##      'ind' (factor 3): O10, O11, O12.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/slice.html">slice</a></span>(olea, <span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>))</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 110 open outlines (99 +/- 4 coordinates)
##  - $fac: 4 classifiers:
##      'var' (factor 2): MouBo1, PicMa.
##      'domes' (factor 2): cult, wild.
##      'view' (factor 2): VD, VL.
##      'ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter: select individual based on a logical condition</span>
<span class="kw"><a href="../reference/filter.html">filter</a></span>(olea, domes<span class="op">==</span><span class="st">"cult"</span>, view<span class="op">!=</span><span class="st">"VD"</span>)</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 60 open outlines (97 +/- 4 coordinates)
##  - $fac: 4 classifiers:
##      'var' (factor 2): Aglan, PicMa.
##      'domes' (factor 1): cult.
##      'view' (factor 1): VL.
##      'ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select: pick, reorder columns from the $fac</span>
<span class="kw"><a href="../reference/select.html">select</a></span>(olea, <span class="dv">1</span>, <span class="dt">Ind=</span>ind)</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 210 open outlines (99 +/- 4 coordinates)
##  - $fac: 2 classifiers:
##      'var' (factor 4): Aglan, Cypre, MouBo1, PicMa.
##      'Ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># rename: rename columns (select can also do it)</span>
<span class="kw"><a href="../reference/rename.html">rename</a></span>(olea, <span class="dt">domesticated=</span>domes)</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 210 open outlines (99 +/- 4 coordinates)
##  - $fac: 4 classifiers:
##      'var' (factor 4): Aglan, Cypre, MouBo1, PicMa.
##      'domesticated' (factor 2): cult, wild.
##      'view' (factor 2): VD, VL.
##      'ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mutate: add new columns</span>
<span class="kw"><a href="../reference/mutate.html">mutate</a></span>(olea, <span class="dt">fake=</span><span class="kw">factor</span>(<span class="kw">rep</span>(letters[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">each=</span><span class="dv">105</span>)))</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 210 open outlines (99 +/- 4 coordinates)
##  - $fac: 5 classifiers:
##      'var' (factor 4): Aglan, Cypre, MouBo1, PicMa.
##      'domes' (factor 2): cult, wild.
##      'view' (factor 2): VD, VL.
##      'ind' (factor 30): O1, O10, O11, O12, O13, O14, O15, O16, O17, O18, O19 ... + 19 more.
##      'fake' (factor 2): a, b.</code></pre>
<p>And you can pipe those operations: say, we only want dorsal views from domesticated individuals, for a (renamed) ‘status’ column, and drop the ‘ind’ column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olea <span class="op">%&gt;%</span><span class="st"> </span>
<span class="kw"><a href="../reference/filter.html">filter</a></span>(domes<span class="op">==</span><span class="st">"cult"</span>, view<span class="op">==</span><span class="st">"VD"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="kw"><a href="../reference/rename.html">rename</a></span>(<span class="dt">domesticated=</span>domes) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="kw"><a href="../reference/select.html">select</a></span>(<span class="op">-</span>ind)</code></pre></div>
<pre><code>## An Opn object with: 
##  - $coo: 90 open outlines (100 +/- 2 coordinates)
##  - $fac: 3 classifiers:
##      'var' (factor 3): Aglan, Cypre, PicMa.
##      'domesticated' (factor 1): cult.
##      'view' (factor 1): VD.</code></pre>
<p>This should save some headaches.</p>
</div>
<div id="new-verbs-ala-dplyr" class="section level3">
<h3 class="hasAnchor">
<a href="#new-verbs-ala-dplyr" class="anchor"></a>New verbs ala <code>dplyr</code>
</h3>
<p>New verbs are implemented: for instance, you can <code>chop</code> (a rougher slicing) according to a condition: this will create a list, on which you can apply further operations, then <code>combine</code> it back. This is particularly useful when you want to apply independant treatments to different partitions, eg orthogonal views of your model. Prior to this, we can use <code>table</code> to cross-tabulate data from <code>$fac</code>. We could have done the first step of what follows with <code>rm_uncomplete</code> that drops (if any) missing data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/table.html">table</a></span>(olea, <span class="st">"view"</span>, <span class="st">"var"</span>) 
<span class="co"># we drop 'Cypre' since there is no VL for 'Cypre' var</span>
olea <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filter.html">filter</a></span>(var <span class="op">!=</span><span class="st"> "Cypre"</span>) <span class="op">%&gt;%</span><span class="st">              </span>
<span class="st">  </span><span class="co"># split, do morphometrics, combine</span>
<span class="st">  </span><span class="kw"><a href="../reference/chop.html">chop</a></span>(view) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">lapply</span>(opoly) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/combine.html">combine</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># note the two views in the morphospace</span>
<span class="st">  </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"var"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-44-1.png" width="672"></p>
<pre><code>##     var
## view Aglan Cypre MouBo1 PicMa
##   VD    30    30     30    30
##   VL    30     0     30    30</code></pre>
<p>Also if you need some <code>group_by</code> operation, you can use the <code>fac</code> <code>data.frame</code>. For instance if you want to <code>arrange</code> based on a column and retain the top 10 shapes, then you can:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">retain &lt;-<span class="st"> </span>x<span class="op">$</span>fac <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># here we sort and create an id</span>
<span class="st">  </span><span class="kw"><a href="../reference/arrange.html">arrange</a></span>(your_column) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mutate.html">mutate</a></span>(<span class="dt">.id=</span><span class="dv">1</span><span class="op">:</span><span class="kw">n</span>()) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(taxa) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/slice.html">slice</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>) <span class="op">%$%</span><span class="st"> </span>.id
<span class="co"># it worked!</span>
x <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/slice.html">slice</a></span>(retain) <span class="op">%$%</span><span class="st"> </span><span class="kw"><a href="../reference/table.html">table</a></span>(fac<span class="op">$</span>taxa)</code></pre></div>
</div>
<div id="various-helpers" class="section level3">
<h3 class="hasAnchor">
<a href="#various-helpers" class="anchor"></a>Various helpers</h3>
<p>Some methods help, on <code>Coe</code> objects to: * select groups with at least a certain number of individuals in them: <code>at_least</code> * removes outliers : <code>which_out</code> * sample a given number: <code>sample_n</code>; * sample a given proportion: <code>sample_frac</code>; * generate new individuals based on calibrated Gaussian coefficient generation: <code>breed</code>; * generate new individuals based on permutations: <code>perm</code>.</p>
<p>Several shortcuts are implemented on <code>Coo</code> and <code>Coe</code> objects: * <code>names</code> returns shape names; * <code>length</code> returns their number; * <code>table</code> cross-tabulates the<code>$fac</code> component; * <code>Ntable</code> does the same job and plots a confusion matrix; * <code>[]</code> extracts one (or more) shape; * <code>$</code> can access either a shape name or a column name for the <code>$fac</code>.</p>
<p>Try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(bot)
<span class="kw">length</span>(bot)
<span class="kw"><a href="../reference/table.html">table</a></span>(olea, <span class="st">"var"</span>, <span class="st">"domes"</span>)
<span class="kw">Ntable</span>(olea, <span class="st">"var"</span>, <span class="st">"domes"</span>)
bot[<span class="dv">1</span>]
bot[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
bot<span class="op">$</span>brahma
bot<span class="op">$</span>type</code></pre></div>
</div>
</div>
<div id="babel-importexport" class="section level2">
<h2 class="hasAnchor">
<a href="#babel-importexport" class="anchor"></a>Babel import/export</h2>
<p>There are various morphometrics formats in the wild, almost as much as softwares. Momocs tries to create bridges between them, all gathered in the <strong>Babel</strong> family.</p>
<div id="bridges-within-r" class="section level3">
<h3 class="hasAnchor">
<a href="#bridges-within-r" class="anchor"></a>Bridges within R</h3>
<p>You can convert from/to array, matrix, list or data.frame with the functions <code>{a, m, l, d}2{a, m, l, d}</code>. For instance, <code>l2a</code> converts a list into an array that you can use with <code>geomorph</code>; <code>a2l</code> does the inverse operation.</p>
<p>Imagine you want to import pupfish from <code>geomorph</code> as a <code>Ldk</code> object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(geomorph)
<span class="kw">data</span>(pupfish)
<span class="kw">str</span>(pupfish)
<span class="co"># so $coords will become $coo, and</span>
<span class="co"># all other components will be turned into a data.frame to feed $fac</span>
<span class="co"># with a single line</span>
<span class="kw"><a href="../reference/Ldk.html">Ldk</a></span>(<span class="dt">coo=</span>pupfish<span class="op">$</span>coords <span class="op">%&gt;%</span><span class="st"> </span>a2l,
<span class="dt">fac=</span>pupfish[<span class="op">-</span><span class="dv">1</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>())</code></pre></div>
</div>
<div id="import-from-stereomorph" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-stereomorph" class="anchor"></a>Import from StereoMorph</h3>
<p>If you use this excellent package by Olsen to digitize landmarks and curves, you can import them, from the files produced with the functions <code>import_StereoMorph_ldk</code> and <code>import_StereoMorph_curve</code>.</p>
</div>
<div id="import-from-tps-and-other-digitizing-softwares" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-tps-and-other-digitizing-softwares" class="anchor"></a>Import from tps and other digitizing softwares</h3>
<ul>
<li>
<code>.tps</code> files can be read with <code>import_tps</code>
</li>
<li>
<code>.nts</code> files an be read with <code>nts2Coo</code> (will be turned into <code>import_nts</code> soon)</li>
</ul>
</div>
<div id="direct-build-of-coe-objects-youre-not-bound-with-momocs-from-the-shapes-step-ie-you-do-not-have-to-start-from-coo-objects--for-instance-if-you-have-a-matrix-of-coefficients-you-can-directly-build-an-outcoe-with-the-builder-see-below--same-approach-for-opncoe-and-tracoe-have-a-look-to-the-help-files-of-these-builders-" class="section level3">
<h3 class="hasAnchor">
<a href="#direct-build-of-coe-objects-youre-not-bound-with-momocs-from-the-shapes-step-ie-you-do-not-have-to-start-from-coo-objects--for-instance-if-you-have-a-matrix-of-coefficients-you-can-directly-build-an-outcoe-with-the-builder-see-below--same-approach-for-opncoe-and-tracoe-have-a-look-to-the-help-files-of-these-builders-" class="anchor"></a>Direct build of <em>Coe objects </em> You’re not bound with Momocs from the “shapes” step, ie you do not have to start from <code>Coo</code> objects. For instance if you have a matrix of coefficients, you can directly build an <code>OutCoe</code> with the builder (see below). Same approach for <code>OpnCoe</code> and <code>TraCoe</code>; have a look to the help files of these builders.</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we simulate an imported matrix of coordinates, eg from a .csv</span>
coeffs_from_the_wild &lt;-<span class="st"> </span>bot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dv">6</span>) <span class="op">%$%</span><span class="st"> </span>coe
coeffs_in_Momocs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/OutCoe.html">OutCoe</a></span>(<span class="dt">coe=</span>coeffs_from_the_wild, <span class="dt">method=</span><span class="st">"efourier"</span>, <span class="dt">norm=</span><span class="ot">TRUE</span>)
coeffs_in_Momocs <span class="op">%&gt;%</span><span class="st"> </span>PCA <span class="op">%&gt;%</span><span class="st"> </span>plot</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-48-1.png" width="672"></p>
</div>
<div id="import-misc" class="section level3">
<h3 class="hasAnchor">
<a href="#import-misc" class="anchor"></a>Import misc</h3>
<ul>
<li>If covariables are encoded in filenames, which is a good practice, eg if you have files named <code>spA_group7_ind4_VL.{txt|jpg|etc.}</code>, use <code>lf_structure</code>;</li>
<li>If you need rewriting rules on your <code>$fac</code>, <code>rw_rule</code> is your friend;</li>
<li>If you need to rescale imported coordinates, see <code>rescale</code>;</li>
<li>If you need to tie images (eg outlines) and <code>.txt</code> (eg coordinates of landmarks on them), see <code>tie_jpg_txt</code>.</li>
</ul>
</div>
<div id="save-from-r" class="section level3">
<h3 class="hasAnchor">
<a href="#save-from-r" class="anchor"></a>Save from R</h3>
<p>The best way to save a Momocs object is probably to use the base <code>save</code> function. You can call it back afterwards with <code>load</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(bot, <span class="dt">file=</span><span class="st">"Bottles.rda"</span>)
<span class="co"># closing R, going to the beach, back at work</span>
<span class="kw">load</span>(<span class="st">"Bottles.rda"</span>) <span class="co"># bot is back</span></code></pre></div>
</div>
<div id="export-from-r" class="section level3">
<h3 class="hasAnchor">
<a href="#export-from-r" class="anchor"></a>Export from R</h3>
<p>Any Momocs object, <code>Coo</code>s, <code>Coe</code>s, <code>PCA</code>s, etc. can be turned into a <code>data.frame</code> with <code>as_df</code>. Work with <code>dplyr</code>, <code>ggplot2</code> is made easy and you can export it as <code>.txt</code>, <code>.csv</code> “by hand” or use the <code>export</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot <span class="op">%&gt;%</span><span class="st"> </span>as_df <span class="co"># then %&gt;% write.table</span>
bot <span class="op">%&gt;%</span><span class="st"> </span>efourier <span class="op">%&gt;%</span><span class="st"> </span>export
bot <span class="op">%&gt;%</span><span class="st"> </span>efourier <span class="op">%&gt;%</span><span class="st"> </span>PCA <span class="op">%&gt;%</span><span class="st"> </span>export</code></pre></div>
<p>But, of course, you can directly access information within the Momocs objects; try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># from Coo objects</span>
bot<span class="op">$</span>coo <span class="co"># list of matrices (of xy coordinates)</span>
bot<span class="op">$</span>fac <span class="co"># data.frame for covariates</span>

<span class="co"># from Coe objects</span>
bot.f<span class="op">$</span>coe <span class="co"># matrix of coefficients</span>
bot.f<span class="op">$</span>fac <span class="co"># data.frame for covariates</span>
<span class="kw"><a href="../reference/as_df.html">as_df</a></span>(bot.f)

<span class="co"># from PCA objects</span>
bot.p<span class="op">$</span>x        <span class="co"># scores</span>
bot.p<span class="op">$</span>rotation <span class="co"># rotation matrix</span></code></pre></div>
</div>
</div>
<div id="graphics" class="section level2">
<h2 class="hasAnchor">
<a href="#graphics" class="anchor"></a>Graphics</h2>
<p>The most common graphics have been already presented in previous sections. But there are many more of them, and they are gathered in the <strong>Graphics</strong> family that can be accessed with … TODO.</p>
<div id="tps_-thin-plate-splines" class="section level3">
<h3 class="hasAnchor">
<a href="#tps_-thin-plate-splines" class="anchor"></a><code>tps_*</code>: Thin Plate Splines</h3>
<p>TPS have not been presented before but here there are:</p>
<pre><code><a href="../reference/tps_grid.html">tps_grid(beer, whisky)
tps_arr(beer, whisky)
tps_iso(beer, whisky)</a></code></pre>
<p>Again, plenty options in <code>?tps_*</code>.</p>
<p>You may also like lolliplots and friends:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_lolli.html">coo_lolli</a></span>(beer, whisky); <span class="kw">title</span>(<span class="st">"coo_lolli"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-52-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_arrows.html">coo_arrows</a></span>(beer, whisky); <span class="kw">title</span>(<span class="st">"coo_arrow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-52-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># an example with coo_ruban</span>
<span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(beer) <span class="co"># to get the first plot </span>
<span class="kw"><a href="../reference/coo_ruban.html">coo_ruban</a></span>(beer, <span class="kw"><a href="../reference/edm.html">edm</a></span>(beer, whisky), <span class="dt">lwd=</span><span class="dv">8</span>) <span class="co"># we add ruban based from deviations</span>
<span class="kw"><a href="../reference/coo_draw.html">coo_draw</a></span>(whisky)
<span class="kw">title</span>(<span class="st">"coo_ruban"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-52-3.png" width="672"></p>
</div>
<div id="ggplot2" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2" class="anchor"></a><code>ggplot2</code>
</h3>
<p>Many graphical function have their <code>ggplot2</code> counterpart:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(bot)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-53-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/panel2.html">panel2</a></span>(<span class="kw"><a href="../reference/slice.html">slice</a></span>(bot, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)) <span class="co"># for speed sake</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-53-2.png" width="672"></p>
<p>And the classical <code>ggplot2</code> grammar applies:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
gg &lt;-<span class="st"> </span><span class="kw"><a href="../reference/panel2.html">panel2</a></span>(<span class="kw"><a href="../reference/slice.html">slice</a></span>(bot, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>)) <span class="co"># for speed sake</span>
gg <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_minimal</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-54-1.png" width="672"></p>
<p>Otherwise, you can build your own plots. Let’s begin with a simple example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we build a ggplot object from a shape turned into a data.frame</span>
shapes[<span class="dv">4</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/bridges.html">m2d</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(x, y) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_path</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/coord_fixed">coord_equal</a></span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/labs">labs</a></span>(<span class="dt">title=</span><span class="st">"ggplot2 Meow"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_minimal</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-55-1.png" width="672"></p>
<p>Here is a more complicated plot that takes profit of <code>as_df</code> methods that turns objects into <code>ggplot2</code>-friendly <code>data.frame</code>s:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/as_df.html">as_df</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>() <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x=</span>PC1, <span class="dt">y=</span>PC2, <span class="dt">col=</span>type) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/coord_fixed">coord_equal</a></span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_density_2d">geom_density2d</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_light</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-56-1.png" width="672"></p>
</div>
<div id="color-palettes" class="section level3">
<h3 class="hasAnchor">
<a href="#color-palettes" class="anchor"></a>Color palettes</h3>
<ul>
<li>Nice palettes are bundled with Momocs, see <code><a href="../reference/color_palettes.html">?Palettes</a></code>.</li>
<li>Overall, <a href="https://CRAN.R-project.org/package=RColorBrewer">RColorBrewer</a> is more than an option if you work with colors.</li>
</ul>
</div>
</div>
<div id="frequently-asked-questions" class="section level2">
<h2 class="hasAnchor">
<a href="#frequently-asked-questions" class="anchor"></a>Frequently asked questions</h2>
<ul>
<li>
<strong>How to access all methods for a given class?</strong>
<ul>
<li>
<code>methods(class="your_class")</code>, eg <code>methods(class="Opn")</code>
</li>
</ul>
</li>
<li>
<strong>How to access methods defined for a particular function?</strong>
<ul>
<li>
<code>methods("your_function")</code>, eg <code>methods("fgProcrustes")</code>
</li>
</ul>
</li>
<li>
<strong>How can I access help for a particular methods?</strong>
<ul>
<li>
<code>function.class</code>, eg <code><a href="../reference/plot.PCA.html">?plot.PCA</a></code>
</li>
</ul>
</li>
<li>
<strong>How can I access the code of a method?</strong>
<ul>
<li>For functions, just type its name, eg <code>lf_structure</code>; for methods add its class, eg <code>plot.PCA</code>
</li>
</ul>
</li>
<li><strong>To be continued</strong></li>
</ul>
<hr>
<ul>
<li>Done with love by Vincent Bonhomme <code>&lt;bonhomme.vincent@gmail.com&gt;</code>.</li>
<li>Help/suggestions/review/critics are welcome.</li>
<li>Sources are <a href="https://github.com/MomX/Momocs/tree/master/vignettes">there</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#preliminaries">Preliminaries</a></li>
      <li><a href="#design">Design</a></li>
      <li><a href="#single-shapes">Single shapes</a></li>
      <li><a href="#morphometrics">Morphometrics</a></li>
      <li><a href="#multivariate-statistics">Multivariate statistics</a></li>
      <li><a href="#manipulating-objects">Manipulating objects</a></li>
      <li><a href="#babel-importexport">Babel import/export</a></li>
      <li><a href="#graphics">Graphics</a></li>
      <li><a href="#frequently-asked-questions">Frequently asked questions</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://www.vincentbonhomme.fr">Vincent Bonhomme</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
