<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Momocs speed dating • Momocs</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Momocs</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Momocs_grindr.html">Layered shape and multivariate plots with grindr</a>
    </li>
    <li>
      <a href="../articles/Momocs_speed_dating.html">Momocs speed dating</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MomX/Momocs">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Momocs speed dating</h1>
                        <h4 class="author">Vincent Bonhomme</h4>
            
            <h4 class="date">2018-04-09</h4>
          </div>

    
    
<div class="contents">
<p>Note that this vignette is, as well as Momocs, being entirely rewritten this day. As per request, I put it back in the package but things may evolve on a weekly basis until Momocs 2.0</p>
<div id="preliminaries" class="section level2">
<h2 class="hasAnchor">
<a href="#preliminaries" class="anchor"></a>Preliminaries</h2>
<div id="abstract" class="section level3">
<h3 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h3>
<p>Momocs aims to provide a complete and convenient toolkit for morphometrics. It is intended for scientists interested in describing quantitatively the shape, and its (co)variations, of the objects they study.</p>
<p>In the last decade, R has become the open-source <em>lingua franca</em> for statistics, and morphometrics known its so-called “revolution”. Nevertheless, morphometric analyses still have to be carried out using various software packages either dedicated to a particular morphometric and/or for which source code is mostly unavailable and/or copyrighted. Moreover, most of existing software packages cannot be extended and their bugs are hard to detect and thus correct. This situation is detrimental to morphometrics: time is wasted, analyses are restricted to available methods, and last but not least, are poorly reproducible. This impedes collaborative effort both in software development and in morphometric studies.</p>
<p>By gathering the common morphometric approaches in an open-source environment and welcoming contributions, Momocs is an attempt to solve this twofold problem.</p>
<p>Momocs hinges on the core functions published in the must-have <em><a href="http://www.springer.com/us/book/9780387777894">Morphometrics with R</a></em> by <a href="http://www.isem.univ-montp2.fr/recherche/equipes/macroevolution-et-developpement/personnel/claude-julien/">Julien Claude</a> (2008), but has been further extended to allow other shape description systems. So far, configurations of landmarks, outlines and open outline analyses, along with some facilities for traditional morphometrics are implemented.</p>
<p>Prior to analysis, Momocs can be used to acquire and manipulate data or to import/export from/to other formats. Momocs also has the facility for a wide range of multivariate analyses and production of the companion graphics. Thus a researcher will find that just a few lines of code will provide initial results, but the methods implemented can be finely tuned and extended according to the user’s needs.</p>
<ul>
<li>If you use it, cite it: <code>citation("Momocs")</code>.</li>
<li>This citation refers to an obsolete version of Momocs, only handling outline analyses. The next companion and seminal paper is submitted.</li>
</ul>
</div>
<div id="survival-tips" class="section level3">
<h3 class="hasAnchor">
<a href="#survival-tips" class="anchor"></a>Survival tips</h3>
<ul>
<li>This vignette introduces Momocs; more specific help can be find in function’s help files, for instance <code><a href="../reference/efourier.html">?efourier</a></code>.</li>
<li>There is a much nicer <a href="http://momx.github.io/Momocs/">online version of this manual</a> that can be accessed from the console with, e.g. <code><a href="../reference/Momocs_help.html">Momocs_help("efourier")</a></code>.</li>
<li>Feel free to contribute to Momocs through <a href="http://www.github.com/MomX/Momocs">GitHub</a>: report issues, ask for new features, share data and methods, correct typos, write better vignettes, helpfiles, or whatever pleases you. If you have never heard of GitHub, that’s definitely worth <a href="http://r-pkgs.had.co.nz/git.html">a look</a>.</li>
<li>Feel free to drop me a line, should you need a hand or would like to collaborate with me: <code>bonhomme.vincent@gmail.com</code>.</li>
</ul>
</div>
<div id="get-install-and-use-it" class="section level3">
<h3 class="hasAnchor">
<a href="#get-install-and-use-it" class="anchor"></a>Get, install and use it</h3>
<p>First, of all, let’s download the last version of Momocs. You will need to install the <code>devtools</code> package to get it from my <a href="http://www.github.com/MomX/Momocs">GitHub</a> repository :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"MomX/Momocs"</span>)</code></pre></div>
<p>The typical <code>install_packages("Momocs")</code> will get you the last CRAN version of Momocs, but the GitHub version is preferred as Momocs is still under active development.</p>
<p>We can start using Momocs, as long as it has been loaded using:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Momocs)</code></pre></div>
</div>
</div>
<div id="design" class="section level2">
<h2 class="hasAnchor">
<a href="#design" class="anchor"></a>Design</h2>
<p>Keywords used all accross Momocs are introduced here in <strong>bold</strong>.</p>
<p>Morphometrics is the ugly job of turning beautiful shapes into quantitative variables. Just kidding, that’s pretty exciting.</p>
<p>A <strong>shape</strong> is defined as a collection of <code>(x; y)</code> coordinates. No 3D yet but different <strong>families</strong> can be handled: <strong>outlines</strong>, here in a first-quarter moon ; <strong>open outlines</strong>, here is the sterile valve of an olive stone; <strong>configuration of landmarks</strong>; here, hologous points from a mosquito wing.</p>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-4-1.png" width="672"><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-4-2.png" width="672"><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-4-3.png" width="672"></p>
<p>They are all single shapes defined by a matrix of <code>(x; y)</code> coordinates; here are the first points of the moon:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shapes[<span class="dv">18</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()</code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]  200   50
## [2,]  199   49
## [3,]  198   49
## [4,]  197   50
## [5,]  196   50
## [6,]  195   49</code></pre>
<p>A few dozens of operations on single shapes are implemented such as: plotting, centering, calculating areas, etc. These 70+ operations can be accessed with <code>apropos("coo_")</code>. But working on single shapes is quite boring.</p>
<p>Shapes can be organized into a <strong>collection</strong> of coordinates: a <code>Coo</code> object that carries:</p>
<ul>
<li>a component named <code>$coo</code>, a <code>list</code> of shapes (as <code>matrix</code>.ces);</li>
<li>most of the time, a component named <code>$fac</code>, a <code>data.frame</code> to store covariates, either <code>factor</code>s or <code>numeric</code>s;</li>
<li>possibly, other components of interest.</li>
</ul>
<p>One can do several things with a <code>Coo</code> object: visualize it, apply morphometric operations, handle the data it contains, but in the end, a __ morphometric method__ will turn coordinates into coefficients.</p>
<p>Such morphometric operation on coordinates produce a collection of coefficients: a <code>Coe</code> object that carries:</p>
<ul>
<li>a component named <code>$coe</code>, a <code>matrix</code> of coefficients;</li>
<li>if present in <code>Coo</code>, <code>$fac</code> is inherited;</li>
<li>possibly, other components of interest.</li>
</ul>
<p>This can be summarized as follows:</p>
<table class="table">
<thead><tr class="header">
<th><code>Coo</code></th>
<th>+</th>
<th>Morphometric method</th>
<th>=</th>
<th><code>Coe</code></th>
</tr></thead>
<tbody><tr class="odd">
<td>
<code>(x; y)</code> coordinates</td>
<td>+</td>
<td>appropriate method</td>
<td>=</td>
<td>quantitative variables</td>
</tr></tbody>
</table>
<p><code>Coo</code> objects are collections of coordinates that become <code>Coe</code> objects when an appropriate morphometric method is applied on them.</p>
<p>Some operations on <code>Coo</code>/<code>Coe</code> are <strong>generic</strong> in that they do not depend of the nature of the shape. For instance, centering a configuration of landmarks or an outline, or calculating their centroid size is, mathematically, the same <strong>generic</strong> operation. But some operations on shapes are <strong>specific</strong> to a peculiar family. For instance, calculating elliptical Fourier transforms on a configuration of landmarks would make no sense.</p>
<p>Momocs implement this desirable behavior and defines <strong>classes</strong> and <strong>subclasses</strong>, as S3 objects.</p>
<table class="table">
<thead><tr class="header">
<th><code>Coo</code></th>
<th>Morphometrics methods</th>
<th><code>Coe</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>
<code>OutCoo</code> (outlines)</td>
<td>
<code>efourier</code>, <code>rfourier</code>, <code>sfourier</code>, <code>tfourier</code>
</td>
<td><code>OutCoe</code></td>
</tr>
<tr class="even">
<td>
<code>OpnCoo</code> (open outlines)</td>
<td>
<code>npoly</code>, <code>opoly</code>, <code>dfourier</code>
</td>
<td><code>OpnCoe</code></td>
</tr>
<tr class="odd">
<td>
<code>LdkCoo</code> (configuration of landmarks)</td>
<td>
<code>fgProcrustes</code>, <code>slide</code>
</td>
<td><code>LdkCoe</code></td>
</tr>
</tbody>
</table>
<p>In other words:</p>
<ul>
<li>any collection of shapes belongs to (pick one) <code>{OutCoo, OpnCoo, LdkCoo}</code> and is also a <code>Coo</code> object;</li>
<li>generic and specific methods can be applied to it</li>
<li>a collection of coefficients is obtain and belongs to (pick one) <code>{OutCoe, OpnCoe, LdkCoe}</code> and is also a <code>Coe</code> object.</li>
</ul>
<p>Finally, generic and specific operations can be applied to the <code>Coe</code> objects, chiefly multivariate methods, capitalicized: <code>PCA</code>, <code>LDA</code>, <code>CLUST</code>, <code>MANOVA</code>(and MANCOVA), <code>MSHAPES</code>, <code>KMEANS</code>, etc.</p>
<p>Overall, Momocs implements a simple and consistent grammar that is detailed below. Also, if you’re familiar with modern R and the Hadley-verse, you should feel home as <code>ggplot2</code> graphics, <code>dplyr</code> verbs and <code>magrittr</code> pipes are implemented.</p>
</div>
<div id="single-shapes" class="section level2">
<h2 class="hasAnchor">
<a href="#single-shapes" class="anchor"></a>Single shapes</h2>
<p>Let’s load one of the Momocs datasets, some various outlines (an <code>Out</code> object):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shapes                    <span class="co"># prints a brief summary</span>
<span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(shapes, <span class="dt">names=</span><span class="ot">TRUE</span>) <span class="co"># base graphics</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<pre><code>## Out (outlines)
##   - 30 outlines, 836 +/- 255 coords (in $coo)
##   - 0 classifiers (in $fac): 
## # A tibble: 0 x 0
##   - also: $ldk</code></pre>
<p><code>shapes</code> is one of the datasets bundled with Momocs. It’s (“lazy”) loaded in memory as soon as you call it, no need for <code>data(shapes)</code>. To see all Momocs’ datasets, try <code>data(package="Momocs")</code>. These datasets are all <code>Coo</code> objects (try <code>class(bot)</code>), ie collection of shapes.</p>
<p>One can do many things on a <code>Coo</code> object, as above, e.g. printing a summary of it (just by typing its name in the console), plotting a family picture with <code>panel</code> or <code>panel2</code>. Note the <code>2</code> that refers to a <code>ggplot2</code> variant of a given plot.</p>
<p>So far, we’re interested in single shapes so let’s extract the 4th shape from <code>shapes</code>, using the traditional syntax. We plot it with <code>coo_plot</code> that comes with several options for plotting all families of shapes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shp &lt;-<span class="st"> </span>shapes[<span class="dv">4</span>]
<span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(shp)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># coo_plot is the base plotter for shapes</span>
<span class="co"># but it can be finely customized, see ?coo_plot</span>
<span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(shp, <span class="dt">col=</span><span class="st">"grey80"</span>, <span class="dt">border=</span><span class="ot">NA</span>, <span class="dt">centroid=</span><span class="ot">FALSE</span>, <span class="dt">main=</span><span class="st">"Meow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-7-2.png" width="672"></p>
<p>Let’s now do some basic operations on this shape. They all named <code>coo_*</code> and you can have the full list with <code>apropos("coo_")</code>. <code>coo_*</code> family encompasses:</p>
<ul>
<li>geometric operations (such as centering, scaling, etc.)</li>
<li>plotting functions</li>
<li>scalar descriptors of shape (such as area, perimeter, circularity, rectilinearity, etc.)</li>
<li>various other operations on a single shape.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(<span class="kw"><a href="../reference/coo_center.html">coo_center</a></span>(shp), <span class="dt">main=</span><span class="st">"centered Meow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(<span class="kw"><a href="../reference/coo_sample.html">coo_sample</a></span>(shp, <span class="dv">64</span>), <span class="dt">points=</span><span class="ot">TRUE</span>, <span class="dt">pch=</span><span class="dv">20</span>, <span class="dt">main=</span><span class="st">"64-pts Meow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-8-2.png" width="672"></p>
<p>Momocs makes use of <a href="https://CRAN.R-project.org/package=magrittr">maggritr</a>’s pipe operators. A nice introduction can be found <a href="https://CRAN.R-project.org/package=magrittr">there</a>. <code>magrittr</code> requires a (very small) cerebral gymnastics at the beginning but the benefits are huge, for defining moprhometric pipelines in Momocs but also for R as a whole. It makes things clearer, it: saves typing; reduces intermediate variable assignation; reads from left to right; substantiates the pipe we (should) have in mind. <code>magrittr</code>’s pipes are already loaded with Momocs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">shapes[<span class="dv">4</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_smooth.html">coo_smooth</a></span>(<span class="dv">5</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_sample.html">coo_sample</a></span>(<span class="dv">64</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">coo_scale</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># pipes can be turned into custom function</span>
cs64 &lt;-<span class="st"> </span><span class="cf">function</span>(x) x <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_sample.html">coo_sample</a></span>(<span class="dv">64</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">coo_scale</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_center.html">coo_center</a></span>()
shapes[<span class="dv">4</span>] <span class="op">%&gt;%</span><span class="st"> </span>cs64 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>() <span class="co"># note the axes</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-9-2.png" width="672"></p>
<p>The most familiar operation can directly be applied on <code>Coo</code> objects:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>coo_center <span class="op">%&gt;%</span><span class="st"> </span>coo_scale <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/coo_alignxax.html">coo_alignxax</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_slidedirection.html">coo_slidedirection</a></span>(<span class="st">"up"</span>) <span class="op">%T&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw">print</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/stack.Coo.html">stack</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<pre><code>## Out (outlines)
##   - 40 outlines, 162 +/- 21 coords (in $coo)
##   - 2 classifiers (in $fac): 
## # A tibble: 40 x 2
##   type   fake 
##   &lt;fct&gt;  &lt;fct&gt;
## 1 whisky a    
## 2 whisky a    
## 3 whisky a    
## 4 whisky a    
## 5 whisky a    
## 6 whisky a    
## # ... with 34 more rows
##   - also: $ldk</code></pre>
</div>
<div id="morphometrics" class="section level2">
<h2 class="hasAnchor">
<a href="#morphometrics" class="anchor"></a>Morphometrics</h2>
<div id="outline-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#outline-analysis" class="anchor"></a>Outline analysis</h3>
<p>A word about data import: you can extract outlines from a list of black masks over a white background, as <code>.jpg</code> images with <code>import_jpg</code>. Have a look to helpfiles (<code>import_jpg</code> and <code>import_jpg1</code>) for more details. Here we do not bother with import since we will use the <code>bot</code>tles outlines dataset bundled with Momocs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(bot)
bot
<span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(bot, <span class="dt">fac=</span><span class="st">"type"</span>, <span class="dt">names=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/stack.Coo.html">stack</a></span>(bot)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-11-2.png" width="672"></p>
<pre><code>## Out (outlines)
##   - 40 outlines, 162 +/- 21 coords (in $coo)
##   - 2 classifiers (in $fac): 
## # A tibble: 40 x 2
##   type   fake 
##   &lt;fct&gt;  &lt;fct&gt;
## 1 whisky a    
## 2 whisky a    
## 3 whisky a    
## 4 whisky a    
## 5 whisky a    
## 6 whisky a    
## # ... with 34 more rows
##   - also: $ldk</code></pre>
<p>Here we will illustrate outline analysis with some elliptical Fourier transforms (but the less used - and tested - radii variation Fourier transforms, its variant used by Renaud et al., and tangent angle Fourier transforms are also implemented with <code>rfourier</code>, <code>sfourier</code> and <code>tfourier</code> respectively).</p>
<p>The idea behind elliptical Fourier transforms is to fit the x and y coordinates separately, that is the blue and red curves below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_oscillo.html">coo_oscillo</a></span>(bot[<span class="dv">1</span>], <span class="st">"efourier"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<pre><code>## # A tibble: 138 x 2
##       dx    dy
##    &lt;dbl&gt; &lt;dbl&gt;
##  1    0.    0.
##  2    3.  -21.
##  3    3.  -32.
##  4    6.  -53.
##  5    9.  -74.
##  6   11.  -84.
##  7   15. -105.
##  8   17. -126.
##  9   20. -147.
## 10   22. -158.
## # ... with 128 more rows</code></pre>
<p>Graphically, this is equivalent to fitting Ptolemaic ellipses on the plane:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/Ptolemy.html">Ptolemy</a></span>(bot[<span class="dv">1</span>])</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>Let’s calibrate the number of harmonics required. More details can be found in their respective help files.</p>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-14-1.png" width="672"><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-14-2.png" width="672"><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-14-3.png" width="672"></p>
<pre><code>## $gg
## 
## $q
##                       h1       h2       h3       h4       h5       h6
## brahma         12.015183 75.77538 84.07922 96.10181 97.34168 98.77976
## caney          25.287301 83.68491 85.38833 97.47106 97.79524 98.32091
## chimay         43.421561 81.54716 86.94665 94.96776 97.39247 98.02442
## corona         18.705505 84.76069 86.29248 96.75231 97.17090 97.83694
## deusventrue    49.960794 89.62993 91.50776 97.37525 98.13378 98.78278
## duvel          43.733170 79.77505 88.14345 93.10172 96.38315 96.68583
## franziskaner   19.307641 80.48552 86.97152 95.60206 96.27931 98.23548
## grimbergen     38.816119 85.20063 87.51264 96.99263 97.57349 97.77877
## guiness        33.238446 80.80898 85.81759 96.51499 97.51998 97.97861
## hoegardeen     25.800395 83.96335 85.84893 97.06527 97.52610 98.40888
## jupiler        24.312112 84.66358 86.03794 97.54700 97.85975 98.70747
## kingfisher     26.135594 84.48777 88.04188 97.17552 97.65600 98.55744
## latrappe       40.006966 67.94796 83.40047 87.90448 94.38216 96.89498
## lindemanskriek 33.668625 84.76379 87.14975 97.23196 97.82567 98.57729
## nicechouffe    38.595506 84.52704 88.31686 96.60627 97.58948 98.32637
## pecheresse     25.817934 84.80965 85.96765 97.04553 97.40322 98.61866
## sierranevada   32.359763 80.66498 86.63786 95.67303 97.22522 97.73540
## tanglefoot     44.144525 75.09545 85.23159 93.76784 96.84021 97.85135
## tauro          24.638135 84.88122 86.17237 97.58843 97.87680 98.73115
## westmalle      26.559319 84.52893 86.38685 96.87098 97.35465 98.65747
## amrut          25.559647 78.32103 83.94673 95.64072 96.52872 97.78039
## ballantines    29.835518 60.96305 83.41052 85.23844 91.95976 93.61191
## bushmills       6.275178 70.46102 82.38351 88.07200 91.04770 94.39188
## chivas         50.407833 81.48772 87.52331 96.55344 97.37883 97.53340
## dalmore        61.130880 79.96678 88.67289 96.22714 97.97682 98.23593
## famousgrouse   20.254164 74.09413 84.59829 93.32387 95.00350 97.87492
## glendronach    24.283297 78.01075 83.60628 95.69481 96.49098 97.67259
## glenmorangie   25.174627 74.67603 80.26797 95.31801 96.15812 98.17947
## highlandpark   30.650892 64.59229 80.58407 87.31869 92.89584 96.53460
## jackdaniels    32.167532 77.27019 83.98360 94.84082 96.62995 97.58573
## jb             28.090999 78.70290 83.65818 95.95055 96.68674 97.35981
## johnniewalker  14.334818 70.90124 83.04662 89.84837 91.85425 96.43178
## magallan       13.369287 71.36417 80.20244 93.20149 94.18488 98.63628
## makersmark     46.604041 85.49398 89.71249 97.23672 97.66333 97.66690
## oban           25.431707 77.83107 82.71750 95.88543 96.59064 97.73377
## oldpotrero     56.238626 88.45486 91.39008 97.95170 98.43647 98.44377
## redbreast      52.375446 82.51750 88.28618 96.97053 97.61586 97.76857
## tamdhu         20.682047 76.19007 81.64790 95.24572 96.18858 97.95112
## wildturkey     34.233782 75.78416 82.54975 95.65258 96.85279 97.66148
## yoichi         26.217978 66.92082 79.80949 88.49538 92.83553 97.35722
##                      h7       h8       h9      h10      h11      h12
## brahma         98.93834 98.96885 99.35135 99.49083 99.78892 99.83907
## caney          98.58170 98.93137 99.66487 99.76105 99.83464 99.83984
## chimay         98.84439 98.90095 99.68603 99.70994 99.75952 99.76441
## corona         98.18097 98.99520 99.49044 99.56051 99.63586 99.64366
## deusventrue    99.43345 99.48470 99.74104 99.74521 99.88867 99.88932
## duvel          98.45994 98.63711 99.66225 99.68683 99.73484 99.80925
## franziskaner   98.66182 98.93920 99.37315 99.44664 99.73243 99.74755
## grimbergen     98.37062 98.76975 99.70911 99.78080 99.79771 99.81567
## guiness        98.35037 98.36662 99.44825 99.59655 99.80284 99.80681
## hoegardeen     98.75081 99.09224 99.51903 99.59162 99.77548 99.79636
## jupiler        98.99411 99.33156 99.71026 99.75434 99.88047 99.89240
## kingfisher     98.88940 99.15843 99.51806 99.56392 99.74486 99.77429
## latrappe       99.16748 99.31723 99.59669 99.61555 99.69219 99.74053
## lindemanskriek 99.04016 99.22543 99.60968 99.61839 99.85332 99.86514
## nicechouffe    99.00682 99.12873 99.60201 99.61001 99.85551 99.86228
## pecheresse     99.00885 99.37037 99.68087 99.71826 99.88329 99.89623
## sierranevada   98.21142 98.22338 99.32287 99.52992 99.83491 99.84351
## tanglefoot     98.23226 98.32455 99.16811 99.30599 99.69542 99.70261
## tauro          99.00773 99.33870 99.70904 99.74972 99.87046 99.88265
## westmalle      99.08123 99.40464 99.64370 99.64988 99.78594 99.79494
## amrut          98.21574 98.23525 98.95793 99.26041 99.76412 99.79776
## ballantines    98.37535 99.05510 99.58352 99.62876 99.63775 99.69544
## bushmills      96.39559 98.10980 99.27995 99.46343 99.59144 99.73298
## chivas         98.27691 98.49702 99.73432 99.75594 99.79557 99.82221
## dalmore        98.74891 98.83402 99.56825 99.64969 99.73304 99.83380
## famousgrouse   98.37369 98.39547 98.80247 98.82414 99.53318 99.60669
## glendronach    98.07677 98.09667 98.84364 99.22981 99.75547 99.78792
## glenmorangie   98.56749 98.62995 99.14052 99.19841 99.68853 99.70351
## highlandpark   98.58425 98.91184 98.96625 99.09444 99.36492 99.57817
## jackdaniels    97.96643 98.00015 98.85683 99.10608 99.76384 99.81473
## jb             97.88849 97.90482 98.88380 99.28781 99.67140 99.76931
## johnniewalker  98.10439 98.90945 99.27686 99.34737 99.55178 99.55770
## magallan       98.69120 98.78238 99.00320 99.12456 99.52406 99.62456
## makersmark     98.97988 99.46612 99.68040 99.74369 99.79648 99.81828
## oban           98.20060 98.26030 98.94173 99.35375 99.76311 99.79074
## oldpotrero     99.35782 99.46957 99.76279 99.76400 99.88371 99.90242
## redbreast      98.46627 98.64025 99.75810 99.79416 99.84011 99.86920
## tamdhu         98.28202 98.28570 98.82264 99.06467 99.74465 99.77447
## wildturkey     98.18703 98.20022 99.19434 99.33499 99.68401 99.75038
## yoichi         98.69958 99.04137 99.12485 99.22615 99.52969 99.57055
##                     h13      h14      h15      h16      h17      h18
## brahma         99.89334 99.90228 99.92438 99.93392 99.94305 99.94815
## caney          99.91300 99.92991 99.93952 99.94303 99.94993 99.95194
## chimay         99.79780 99.82192 99.87449 99.87713 99.88818 99.89764
## corona         99.84697 99.87146 99.91108 99.91738 99.92566 99.92689
## deusventrue    99.93108 99.93449 99.94456 99.95016 99.95601 99.96269
## duvel          99.85724 99.91081 99.92214 99.93122 99.93620 99.94179
## franziskaner   99.85385 99.86304 99.90329 99.90716 99.91978 99.92423
## grimbergen     99.83514 99.85438 99.88190 99.88569 99.89992 99.90247
## guiness        99.82500 99.86587 99.88936 99.90329 99.92157 99.92316
## hoegardeen     99.88952 99.89441 99.91226 99.91365 99.92320 99.92380
## jupiler        99.95425 99.95666 99.96259 99.96330 99.96568 99.96621
## kingfisher     99.88666 99.89972 99.93141 99.93417 99.94411 99.94992
## latrappe       99.79188 99.80657 99.85719 99.87227 99.89659 99.92557
## lindemanskriek 99.93206 99.93307 99.94728 99.95686 99.96300 99.96480
## nicechouffe    99.93681 99.94568 99.95852 99.96538 99.96788 99.97204
## pecheresse     99.93207 99.93876 99.94317 99.94826 99.95274 99.95534
## sierranevada   99.85162 99.87017 99.88954 99.90430 99.93555 99.93935
## tanglefoot     99.77601 99.80763 99.83545 99.85607 99.88003 99.89716
## tauro          99.94771 99.94963 99.95414 99.95456 99.95686 99.95726
## westmalle      99.91248 99.91546 99.92866 99.93556 99.93992 99.94092
## amrut          99.80847 99.83077 99.87500 99.90215 99.93188 99.94105
## ballantines    99.73486 99.82345 99.84452 99.85982 99.87760 99.90773
## bushmills      99.76233 99.78135 99.82564 99.84973 99.87725 99.89756
## chivas         99.82889 99.86325 99.88374 99.93583 99.94196 99.94632
## dalmore        99.84596 99.89710 99.90085 99.92177 99.93113 99.95101
## famousgrouse   99.75558 99.79026 99.80261 99.81449 99.81696 99.85928
## glendronach    99.79455 99.82900 99.86438 99.88246 99.91374 99.92210
## glenmorangie   99.86390 99.86457 99.90075 99.90508 99.91028 99.93101
## highlandpark   99.72174 99.76571 99.88672 99.89054 99.92775 99.92833
## jackdaniels    99.82818 99.84665 99.85012 99.86994 99.90744 99.91330
## jb             99.78217 99.78413 99.86182 99.89958 99.92360 99.92866
## johnniewalker  99.75291 99.75819 99.79936 99.80181 99.83037 99.83962
## magallan       99.80739 99.81897 99.83110 99.84203 99.88210 99.90377
## makersmark     99.88175 99.88608 99.90279 99.90528 99.91975 99.92347
## oban           99.81862 99.82757 99.91184 99.91953 99.95028 99.95075
## oldpotrero     99.91354 99.93543 99.94084 99.95120 99.95568 99.96102
## redbreast      99.87192 99.91377 99.93035 99.95876 99.96089 99.96404
## tamdhu         99.83549 99.85035 99.88252 99.89613 99.91034 99.93325
## wildturkey     99.77506 99.79678 99.84515 99.87528 99.90548 99.92326
## yoichi         99.77068 99.79424 99.87221 99.87420 99.91975 99.92197
##                     h19      h20      h21      h22      h23      h24
## brahma         99.95550 99.95746 99.96417 99.96677 99.97405 99.97863
## caney          99.95622 99.95888 99.97058 99.97227 99.98302 99.98394
## chimay         99.90339 99.91261 99.91529 99.92077 99.92140 99.93824
## corona         99.93706 99.93981 99.95067 99.95162 99.95776 99.96247
## deusventrue    99.96545 99.97365 99.97462 99.97704 99.97718 99.97991
## duvel          99.94319 99.95228 99.95287 99.96245 99.96949 99.97649
## franziskaner   99.93771 99.93920 99.95090 99.95233 99.95949 99.96093
## grimbergen     99.91177 99.91870 99.92329 99.92642 99.93039 99.93874
## guiness        99.93028 99.93387 99.94447 99.95388 99.95922 99.96937
## hoegardeen     99.93408 99.93568 99.94680 99.94756 99.95623 99.95734
## jupiler        99.96928 99.97023 99.97507 99.97576 99.97920 99.97962
## kingfisher     99.95510 99.95766 99.96278 99.96390 99.96758 99.96972
## latrappe       99.93089 99.95401 99.95645 99.96952 99.97292 99.97417
## lindemanskriek 99.97193 99.97255 99.97839 99.97931 99.98097 99.98256
## nicechouffe    99.97455 99.97531 99.97759 99.97826 99.97912 99.98143
## pecheresse     99.96159 99.96453 99.96964 99.97077 99.97388 99.97452
## sierranevada   99.95145 99.95307 99.95565 99.96277 99.96395 99.97039
## tanglefoot     99.89968 99.91328 99.91384 99.92615 99.92758 99.93445
## tauro          99.96042 99.96072 99.96661 99.96707 99.96969 99.97010
## westmalle      99.94628 99.95066 99.95767 99.95913 99.96618 99.96844
## amrut          99.94808 99.95625 99.95838 99.96220 99.96596 99.97068
## ballantines    99.91687 99.93774 99.94053 99.95274 99.95383 99.95853
## bushmills      99.92330 99.93301 99.94578 99.95250 99.95731 99.95771
## chivas         99.94823 99.95561 99.96383 99.97171 99.97307 99.97989
## dalmore        99.95618 99.96249 99.96524 99.96734 99.97189 99.97401
## famousgrouse   99.88900 99.93989 99.95562 99.96013 99.96372 99.96656
## glendronach    99.93080 99.94133 99.94610 99.95415 99.96120 99.96728
## glenmorangie   99.94290 99.95245 99.96035 99.96303 99.96701 99.96979
## highlandpark   99.93284 99.94085 99.94446 99.95644 99.96174 99.96878
## jackdaniels    99.93655 99.93666 99.93968 99.94236 99.94739 99.95498
## jb             99.93439 99.94268 99.95065 99.95835 99.96984 99.97174
## johnniewalker  99.86652 99.89551 99.91658 99.92373 99.93990 99.94002
## magallan       99.94684 99.96455 99.97154 99.97366 99.97655 99.97846
## makersmark     99.93210 99.93672 99.94619 99.94881 99.95083 99.95386
## oban           99.95472 99.95973 99.96159 99.97085 99.97707 99.98017
## oldpotrero     99.96251 99.96731 99.96939 99.97243 99.97489 99.97738
## redbreast      99.96489 99.97017 99.97469 99.97920 99.98006 99.98166
## tamdhu         99.94688 99.94822 99.96482 99.96574 99.97015 99.97341
## wildturkey     99.93224 99.94463 99.94912 99.95786 99.96255 99.96783
## yoichi         99.93503 99.94107 99.94648 99.95962 99.96316 99.97278
##                     h25      h26      h27      h28      h29      h30
## brahma         99.98207 99.98379 99.98470 99.98618 99.98743 99.98885
## caney          99.98473 99.98624 99.98691 99.98869 99.98887 99.98899
## chimay         99.94817 99.95669 99.96457 99.96589 99.97046 99.97151
## corona         99.96717 99.97029 99.97139 99.97479 99.97561 99.97830
## deusventrue    99.98110 99.98466 99.98563 99.98701 99.98718 99.98989
## duvel          99.97868 99.97960 99.98197 99.98235 99.98308 99.98400
## franziskaner   99.96447 99.96642 99.96808 99.96960 99.97048 99.97204
## grimbergen     99.94477 99.95715 99.96154 99.96854 99.97316 99.97565
## guiness        99.97113 99.97550 99.97634 99.97809 99.97878 99.97917
## hoegardeen     99.96016 99.96230 99.96538 99.96926 99.97232 99.97740
## jupiler        99.98138 99.98228 99.98382 99.98570 99.98742 99.98940
## kingfisher     99.97328 99.97507 99.97779 99.98092 99.98257 99.98596
## latrappe       99.97791 99.98003 99.98221 99.98465 99.98616 99.98784
## lindemanskriek 99.98323 99.98523 99.98548 99.98739 99.98797 99.98873
## nicechouffe    99.98186 99.98423 99.98498 99.98594 99.98744 99.98800
## pecheresse     99.97665 99.97827 99.98057 99.98508 99.98655 99.98861
## sierranevada   99.97193 99.97805 99.97818 99.98123 99.98179 99.98365
## tanglefoot     99.93571 99.93984 99.94207 99.94566 99.95188 99.95741
## tauro          99.97194 99.97267 99.97522 99.97828 99.98070 99.98243
## westmalle      99.96859 99.97066 99.97233 99.97904 99.98038 99.98504
## amrut          99.97152 99.97223 99.97503 99.97809 99.97968 99.98394
## ballantines    99.96081 99.96165 99.96382 99.96541 99.97039 99.97149
## bushmills      99.96711 99.97060 99.97825 99.98295 99.98455 99.98585
## chivas         99.98253 99.98375 99.98928 99.98944 99.99277 99.99283
## dalmore        99.97774 99.97781 99.98061 99.98367 99.98534 99.99122
## famousgrouse   99.97346 99.97623 99.97703 99.98009 99.98029 99.98157
## glendronach    99.96972 99.97301 99.98048 99.98455 99.98571 99.98849
## glenmorangie   99.97099 99.97511 99.97882 99.98260 99.98313 99.98525
## highlandpark   99.97404 99.98360 99.98974 99.99045 99.99142 99.99191
## jackdaniels    99.96002 99.96592 99.96796 99.97243 99.97285 99.98266
## jb             99.97436 99.97728 99.97932 99.98125 99.98275 99.98386
## johnniewalker  99.94513 99.94924 99.95376 99.96379 99.96743 99.97928
## magallan       99.97873 99.97982 99.98225 99.98373 99.98606 99.98627
## makersmark     99.95461 99.96330 99.96906 99.97447 99.97682 99.97804
## oban           99.98384 99.98406 99.98506 99.98631 99.98733 99.98999
## oldpotrero     99.97959 99.98149 99.98509 99.98606 99.98781 99.98843
## redbreast      99.98262 99.98289 99.98476 99.98500 99.98921 99.98999
## tamdhu         99.97517 99.98004 99.98048 99.98480 99.98664 99.98887
## wildturkey     99.97109 99.97606 99.97667 99.97997 99.98133 99.98239
## yoichi         99.97394 99.98560 99.98628 99.98973 99.99048 99.99137
##                     h31      h32      h33      h34      h35      h36
## brahma         99.99025 99.99138 99.99249 99.99322 99.99350 99.99410
## caney          99.98914 99.98989 99.99002 99.99074 99.99094 99.99129
## chimay         99.97158 99.97366 99.97544 99.97993 99.98277 99.98620
## corona         99.97887 99.98209 99.98243 99.98382 99.98384 99.98593
## deusventrue    99.99085 99.99211 99.99368 99.99432 99.99473 99.99498
## duvel          99.98507 99.98783 99.99024 99.99294 99.99404 99.99429
## franziskaner   99.97282 99.97480 99.97599 99.97971 99.98128 99.98305
## grimbergen     99.97627 99.97634 99.97664 99.97907 99.98086 99.98373
## guiness        99.97963 99.98072 99.98161 99.98219 99.98261 99.98428
## hoegardeen     99.97864 99.98205 99.98269 99.98424 99.98484 99.98502
## jupiler        99.99025 99.99164 99.99193 99.99219 99.99222 99.99258
## kingfisher     99.98646 99.98755 99.98791 99.98835 99.98860 99.98898
## latrappe       99.98941 99.98972 99.99033 99.99156 99.99172 99.99277
## lindemanskriek 99.98917 99.98967 99.99056 99.99156 99.99218 99.99278
## nicechouffe    99.98900 99.98955 99.99087 99.99115 99.99182 99.99280
## pecheresse     99.98995 99.99004 99.99059 99.99077 99.99095 99.99207
## sierranevada   99.98535 99.98574 99.98595 99.98691 99.98765 99.98785
## tanglefoot     99.96290 99.96650 99.97222 99.97359 99.97595 99.97642
## tauro          99.98342 99.98717 99.98776 99.98792 99.98850 99.98890
## westmalle      99.98561 99.98714 99.98754 99.98957 99.99085 99.99101
## amrut          99.98560 99.98918 99.98937 99.98972 99.99092 99.99148
## ballantines    99.97652 99.97809 99.98278 99.98370 99.98449 99.98557
## bushmills      99.98635 99.98808 99.98868 99.98979 99.99027 99.99114
## chivas         99.99384 99.99410 99.99428 99.99508 99.99529 99.99649
## dalmore        99.99354 99.99561 99.99580 99.99624 99.99644 99.99655
## famousgrouse   99.98338 99.98440 99.98510 99.98683 99.98851 99.98995
## glendronach    99.98887 99.99075 99.99115 99.99210 99.99255 99.99373
## glenmorangie   99.98596 99.98661 99.98814 99.98925 99.98991 99.99078
## highlandpark   99.99262 99.99289 99.99351 99.99389 99.99484 99.99517
## jackdaniels    99.98413 99.98639 99.98680 99.98744 99.98796 99.98913
## jb             99.98463 99.98538 99.98634 99.98750 99.98804 99.98950
## johnniewalker  99.98157 99.98330 99.98378 99.98421 99.98543 99.98563
## magallan       99.98766 99.98830 99.98876 99.98958 99.99092 99.99139
## makersmark     99.97922 99.98341 99.98435 99.98587 99.98684 99.98770
## oban           99.99086 99.99456 99.99468 99.99603 99.99656 99.99736
## oldpotrero     99.99285 99.99378 99.99411 99.99556 99.99570 99.99614
## redbreast      99.99327 99.99361 99.99438 99.99586 99.99592 99.99669
## tamdhu         99.98962 99.98990 99.99035 99.99048 99.99151 99.99234
## wildturkey     99.98442 99.98507 99.99014 99.99131 99.99211 99.99362
## yoichi         99.99406 99.99443 99.99528 99.99556 99.99581 99.99613
##                     h37      h38      h39      h40      h41      h42
## brahma         99.99434 99.99447 99.99475 99.99490 99.99504 99.99523
## caney          99.99147 99.99225 99.99230 99.99266 99.99277 99.99310
## chimay         99.99243 99.99280 99.99333 99.99424 99.99447 99.99517
## corona         99.98614 99.98752 99.98833 99.98963 99.99150 99.99262
## deusventrue    99.99505 99.99532 99.99558 99.99566 99.99573 99.99596
## duvel          99.99438 99.99530 99.99640 99.99756 99.99796 99.99816
## franziskaner   99.98523 99.98718 99.98921 99.99026 99.99113 99.99163
## grimbergen     99.98571 99.98836 99.99109 99.99185 99.99255 99.99327
## guiness        99.98568 99.98756 99.98867 99.98882 99.98950 99.98972
## hoegardeen     99.98544 99.98601 99.98673 99.98958 99.99093 99.99248
## jupiler        99.99295 99.99372 99.99404 99.99543 99.99576 99.99636
## kingfisher     99.98954 99.99060 99.99172 99.99318 99.99359 99.99404
## latrappe       99.99339 99.99452 99.99550 99.99607 99.99650 99.99711
## lindemanskriek 99.99344 99.99413 99.99424 99.99469 99.99525 99.99577
## nicechouffe    99.99336 99.99394 99.99472 99.99515 99.99586 99.99677
## pecheresse     99.99309 99.99420 99.99513 99.99543 99.99550 99.99576
## sierranevada   99.98796 99.98840 99.99003 99.99107 99.99274 99.99330
## tanglefoot     99.97764 99.97778 99.97947 99.98004 99.98363 99.98579
## tauro          99.98929 99.99034 99.99069 99.99271 99.99290 99.99360
## westmalle      99.99135 99.99245 99.99318 99.99381 99.99415 99.99488
## amrut          99.99188 99.99200 99.99286 99.99311 99.99670 99.99743
## ballantines    99.98590 99.98700 99.98746 99.98890 99.98975 99.99072
## bushmills      99.99141 99.99281 99.99395 99.99485 99.99659 99.99673
## chivas         99.99663 99.99671 99.99678 99.99739 99.99747 99.99799
## dalmore        99.99664 99.99686 99.99694 99.99715 99.99718 99.99727
## famousgrouse   99.99178 99.99217 99.99279 99.99321 99.99459 99.99513
## glendronach    99.99415 99.99423 99.99480 99.99507 99.99753 99.99797
## glenmorangie   99.99168 99.99243 99.99338 99.99453 99.99611 99.99632
## highlandpark   99.99601 99.99660 99.99767 99.99815 99.99823 99.99900
## jackdaniels    99.98958 99.99049 99.99218 99.99279 99.99304 99.99406
## jb             99.99077 99.99234 99.99397 99.99486 99.99575 99.99586
## johnniewalker  99.98601 99.98642 99.98697 99.98798 99.98943 99.99109
## magallan       99.99201 99.99290 99.99407 99.99497 99.99531 99.99600
## makersmark     99.98971 99.99118 99.99150 99.99206 99.99301 99.99470
## oban           99.99752 99.99774 99.99792 99.99794 99.99827 99.99831
## oldpotrero     99.99618 99.99624 99.99664 99.99710 99.99732 99.99793
## redbreast      99.99679 99.99686 99.99720 99.99736 99.99745 99.99754
## tamdhu         99.99314 99.99385 99.99449 99.99481 99.99509 99.99537
## wildturkey     99.99448 99.99518 99.99533 99.99580 99.99619 99.99638
## yoichi         99.99643 99.99688 99.99750 99.99761 99.99798 99.99835
##                     h43      h44      h45      h46      h47      h48
## brahma         99.99543 99.99580 99.99620 99.99643 99.99710 99.99755
## caney          99.99397 99.99421 99.99589 99.99692 99.99748 99.99763
## chimay         99.99562 99.99586 99.99697 99.99718 99.99742 99.99763
## corona         99.99341 99.99471 99.99492 99.99532 99.99540 99.99551
## deusventrue    99.99614 99.99633 99.99760 99.99805 99.99873 99.99888
## duvel          99.99850 99.99857 99.99866 99.99874 99.99879 99.99899
## franziskaner   99.99191 99.99238 99.99270 99.99299 99.99332 99.99360
## grimbergen     99.99372 99.99442 99.99514 99.99569 99.99642 99.99679
## guiness        99.98994 99.99040 99.99094 99.99115 99.99225 99.99318
## hoegardeen     99.99358 99.99503 99.99545 99.99577 99.99584 99.99641
## jupiler        99.99721 99.99732 99.99734 99.99740 99.99756 99.99813
## kingfisher     99.99453 99.99463 99.99483 99.99510 99.99523 99.99535
## latrappe       99.99745 99.99753 99.99782 99.99787 99.99833 99.99845
## lindemanskriek 99.99591 99.99610 99.99617 99.99666 99.99712 99.99793
## nicechouffe    99.99728 99.99770 99.99780 99.99813 99.99819 99.99847
## pecheresse     99.99596 99.99612 99.99616 99.99743 99.99774 99.99803
## sierranevada   99.99385 99.99443 99.99478 99.99521 99.99524 99.99535
## tanglefoot     99.98908 99.99290 99.99433 99.99682 99.99711 99.99774
## tauro          99.99425 99.99446 99.99462 99.99499 99.99516 99.99618
## westmalle      99.99559 99.99607 99.99622 99.99655 99.99700 99.99708
## amrut          99.99837 99.99843 99.99859 99.99864 99.99885 99.99893
## ballantines    99.99176 99.99283 99.99362 99.99389 99.99417 99.99439
## bushmills      99.99692 99.99700 99.99715 99.99728 99.99763 99.99828
## chivas         99.99801 99.99825 99.99850 99.99868 99.99879 99.99902
## dalmore        99.99759 99.99802 99.99840 99.99848 99.99860 99.99872
## famousgrouse   99.99582 99.99645 99.99676 99.99710 99.99723 99.99727
## glendronach    99.99866 99.99869 99.99874 99.99882 99.99889 99.99894
## glenmorangie   99.99688 99.99729 99.99749 99.99759 99.99784 99.99788
## highlandpark   99.99905 99.99924 99.99936 99.99949 99.99952 99.99960
## jackdaniels    99.99478 99.99524 99.99528 99.99566 99.99598 99.99630
## jb             99.99626 99.99657 99.99715 99.99717 99.99745 99.99755
## johnniewalker  99.99272 99.99392 99.99514 99.99554 99.99584 99.99658
## magallan       99.99639 99.99672 99.99677 99.99681 99.99717 99.99740
## makersmark     99.99528 99.99564 99.99634 99.99676 99.99683 99.99747
## oban           99.99845 99.99850 99.99854 99.99856 99.99874 99.99886
## oldpotrero     99.99808 99.99860 99.99873 99.99891 99.99895 99.99912
## redbreast      99.99760 99.99765 99.99771 99.99773 99.99800 99.99814
## tamdhu         99.99572 99.99585 99.99627 99.99634 99.99682 99.99764
## wildturkey     99.99665 99.99689 99.99733 99.99738 99.99747 99.99762
## yoichi         99.99851 99.99870 99.99878 99.99883 99.99885 99.99900
##                     h49      h50      h51      h52      h53      h54
## brahma         99.99799 99.99856 99.99911 99.99922 99.99933 99.99964
## caney          99.99793 99.99808 99.99832 99.99866 99.99898 99.99910
## chimay         99.99773 99.99821 99.99841 99.99847 99.99873 99.99887
## corona         99.99599 99.99605 99.99656 99.99677 99.99735 99.99747
## deusventrue    99.99896 99.99910 99.99915 99.99930 99.99936 99.99950
## duvel          99.99909 99.99920 99.99925 99.99936 99.99945 99.99953
## franziskaner   99.99410 99.99514 99.99588 99.99687 99.99787 99.99856
## grimbergen     99.99744 99.99773 99.99790 99.99813 99.99817 99.99828
## guiness        99.99496 99.99626 99.99658 99.99713 99.99797 99.99807
## hoegardeen     99.99675 99.99757 99.99824 99.99864 99.99885 99.99926
## jupiler        99.99841 99.99861 99.99904 99.99921 99.99933 99.99958
## kingfisher     99.99540 99.99563 99.99692 99.99747 99.99763 99.99810
## latrappe       99.99875 99.99890 99.99911 99.99944 99.99961 99.99979
## lindemanskriek 99.99827 99.99875 99.99913 99.99935 99.99951 99.99953
## nicechouffe    99.99864 99.99873 99.99888 99.99903 99.99936 99.99964
## pecheresse     99.99811 99.99823 99.99875 99.99928 99.99944 99.99972
## sierranevada   99.99569 99.99632 99.99685 99.99796 99.99878 99.99924
## tanglefoot     99.99784 99.99785 99.99809 99.99833 99.99864 99.99908
## tauro          99.99692 99.99715 99.99741 99.99764 99.99788 99.99796
## westmalle      99.99772 99.99800 99.99822 99.99886 99.99916 99.99935
## amrut          99.99902 99.99908 99.99936 99.99944 99.99958 99.99961
## ballantines    99.99507 99.99567 99.99722 99.99807 99.99907 99.99947
## bushmills      99.99886 99.99911 99.99941 99.99948 99.99963 99.99965
## chivas         99.99916 99.99925 99.99930 99.99935 99.99943 99.99962
## dalmore        99.99891 99.99903 99.99923 99.99949 99.99962 99.99980
## famousgrouse   99.99739 99.99754 99.99776 99.99784 99.99798 99.99820
## glendronach    99.99926 99.99931 99.99944 99.99948 99.99952 99.99959
## glenmorangie   99.99821 99.99824 99.99872 99.99875 99.99884 99.99895
## highlandpark   99.99966 99.99967 99.99972 99.99973 99.99978 99.99981
## jackdaniels    99.99652 99.99757 99.99775 99.99786 99.99815 99.99833
## jb             99.99788 99.99795 99.99834 99.99849 99.99885 99.99900
## johnniewalker  99.99667 99.99702 99.99753 99.99801 99.99825 99.99875
## magallan       99.99804 99.99822 99.99838 99.99847 99.99852 99.99877
## makersmark     99.99777 99.99850 99.99886 99.99887 99.99903 99.99937
## oban           99.99901 99.99917 99.99934 99.99946 99.99953 99.99965
## oldpotrero     99.99920 99.99925 99.99930 99.99937 99.99950 99.99951
## redbreast      99.99836 99.99838 99.99850 99.99884 99.99915 99.99931
## tamdhu         99.99796 99.99813 99.99864 99.99910 99.99939 99.99940
## wildturkey     99.99776 99.99809 99.99843 99.99859 99.99861 99.99892
## yoichi         99.99903 99.99910 99.99916 99.99921 99.99947 99.99956
##                     h55      h56      h57      h58 h59
## brahma         99.99977 99.99978 99.99980 99.99989 100
## caney          99.99912 99.99947 99.99950 99.99982 100
## chimay         99.99908 99.99936 99.99970 99.99990 100
## corona         99.99835 99.99853 99.99929 99.99967 100
## deusventrue    99.99955 99.99959 99.99975 99.99987 100
## duvel          99.99958 99.99971 99.99983 99.99988 100
## franziskaner   99.99898 99.99921 99.99948 99.99993 100
## grimbergen     99.99930 99.99963 99.99986 99.99992 100
## guiness        99.99828 99.99859 99.99890 99.99925 100
## hoegardeen     99.99970 99.99972 99.99978 99.99991 100
## jupiler        99.99972 99.99978 99.99987 99.99994 100
## kingfisher     99.99840 99.99866 99.99894 99.99954 100
## latrappe       99.99980 99.99981 99.99984 99.99991 100
## lindemanskriek 99.99970 99.99973 99.99978 99.99998 100
## nicechouffe    99.99975 99.99980 99.99985 99.99998 100
## pecheresse     99.99977 99.99980 99.99992 99.99993 100
## sierranevada   99.99951 99.99968 99.99981 99.99984 100
## tanglefoot     99.99918 99.99943 99.99960 99.99971 100
## tauro          99.99811 99.99837 99.99867 99.99912 100
## westmalle      99.99942 99.99946 99.99972 99.99995 100
## amrut          99.99963 99.99973 99.99986 99.99997 100
## ballantines    99.99959 99.99975 99.99981 99.99995 100
## bushmills      99.99967 99.99972 99.99980 99.99984 100
## chivas         99.99967 99.99971 99.99980 99.99988 100
## dalmore        99.99986 99.99989 99.99989 99.99994 100
## famousgrouse   99.99831 99.99854 99.99880 99.99905 100
## glendronach    99.99970 99.99978 99.99983 99.99988 100
## glenmorangie   99.99926 99.99969 99.99970 99.99997 100
## highlandpark   99.99988 99.99993 99.99996 99.99998 100
## jackdaniels    99.99939 99.99956 99.99973 99.99986 100
## jb             99.99933 99.99979 99.99991 99.99998 100
## johnniewalker  99.99883 99.99922 99.99931 99.99951 100
## magallan       99.99886 99.99898 99.99902 99.99955 100
## makersmark     99.99938 99.99960 99.99970 99.99989 100
## oban           99.99983 99.99987 99.99992 99.99995 100
## oldpotrero     99.99961 99.99966 99.99985 99.99990 100
## redbreast      99.99932 99.99941 99.99962 99.99979 100
## tamdhu         99.99973 99.99978 99.99986 99.99995 100
## wildturkey     99.99898 99.99936 99.99964 99.99979 100
## yoichi         99.99970 99.99982 99.99983 99.99998 100
## 
## $minh
##   90%   95%   99% 99.9% 
##     5     5    10    17</code></pre>
<p>Here, 10 harmonics gather 99% of the harmonic power. If you’re happy with this criterium, you can even omit <code>nb.h</code> in <code>efourier</code>: that’s the default parameter, returned with a message.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.f &lt;-<span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(bot, <span class="dt">nb.h=</span><span class="dv">10</span>)
bot.f</code></pre></div>
<pre><code>## An OutCoe object [ elliptical Fourier analysis ]
## --------------------
##  - $coe: 40 outlines described, 10 harmonics
## # A tibble: 40 x 2
##   type   fake 
##   &lt;fct&gt;  &lt;fct&gt;
## 1 whisky a    
## 2 whisky a    
## 3 whisky a    
## 4 whisky a    
## 5 whisky a    
## 6 whisky a    
## # ... with 34 more rows</code></pre>
<p><code>bot.f</code> is a <code>Coe</code> object (and even an <code>OutCoe</code>), you have have a look to the help files to go deeper into Momocs classes.</p>
<p>Let’s have a look to the amplitude of fitted coefficients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(bot.f, <span class="dt">drop=</span><span class="dv">0</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">boxplot</span>(bot.f, <span class="dt">drop=</span><span class="dv">1</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-16-2.png" width="672"></p>
<p>Now, we can calculate a PCA on the <code>Coe</code> object and plot it, along with morphospaces, calculated on the fly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>(bot.f)
<span class="kw">class</span>(bot.p)        <span class="co"># a PCA object, let's plot it</span>
<span class="kw">plot</span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<pre><code>## [1] "PCA"    "prcomp"</code></pre>
<p>Amazing but we will do much better afterwards.</p>
<p>The question of normalization in elliptical Fourier transforms is central. Normalization can either be done beforehand, with geometric operations, or afterhand, directly on the matrix of Fourier coefficients and consuming the first harmonic. In brief, I’m not a big fan of the “use the first harmonics and see what happens” strategy, as some biases can be introduced (and actually quite hard to detect), particularly on rounded/ellipsoid shapes. More can be found in <code><a href="../reference/efourier.html">?efourier</a></code>.</p>
<p>Here is an example of such bias from the <code>molars</code> dataset generously shared by Cornu and Detroit.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># raw molars dataset</span>
<span class="kw"><a href="../reference/stack.Coo.html">stack</a></span>(molars, <span class="dt">title =</span> <span class="st">"Non-aligned molars"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Procrustes-aligned and slided molars</span>
mol.al &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fgProcrustes.html">fgProcrustes</a></span>(molars, <span class="dt">tol =</span> <span class="fl">1e-4</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_slidedirection.html">coo_slidedirection</a></span>(<span class="st">"left"</span>)
<span class="kw"><a href="../reference/stack.Coo.html">stack</a></span>(mol.al, <span class="dt">title=</span><span class="st">"Aligned molars"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Now compare PCA and morphospace using the 1st harmonic alignment</span>
molars <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dt">norm=</span><span class="ot">TRUE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"type"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-3.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># and the a priori normalization </span>
molars <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dt">norm=</span><span class="ot">FALSE</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"type"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-18-4.png" width="672"></p>
<p>Finally, you can drop some harmonics with <code>rm_harm</code>. And methods that removes the bilateral (a)symmetry are implemented: <code>rm_asym</code> and <code>rm_sym</code>, while <code>symmetry</code> calculates some related indices.</p>
</div>
<div id="open-outlines" class="section level3">
<h3 class="hasAnchor">
<a href="#open-outlines" class="anchor"></a>Open outlines</h3>
<p>Open outlines are curves. Methods actually implemented are:</p>
<ul>
<li>
<code>npoly</code> that fit natural polynomials;</li>
<li>
<code>opoly</code> that fit orthogonal (also called Legendre’s) polynomials;</li>
<li>
<code>dfourier</code> for the discrete cosine transform.</li>
</ul>
<p>Note that <code>opoly</code> and <code>npoly</code> can only be used on <em>simple</em> curves, curves that have at most one <code>y</code> for any <code>x</code> coordinates, at least under a given orientation. <code>dfourier</code> can fit <em>complex</em> curves, curves “that back on their feets”.</p>
<p>Here, we will work on the fertile valves of olive stones, a (very partial) dataset provided by my colleagues Terral, Ivorra, and others.</p>
<p>They have two orthogonal views (a lateral and a dorsal view). See the paper cited in <code><a href="../reference/data_olea.html">?olea</a></code> for more details. Let’s explore it a bit:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olea
<span class="kw"><a href="../reference/pile.html">pile</a></span>(olea)     <span class="co"># already aligned \o/</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-19-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/pile.html">pile</a></span>(olea, <span class="op">~</span>view)    <span class="co"># another family picture</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-19-2.png" width="672"></p>
<pre><code>## Opn (curves)
##   - 210 curves, 99 +/- 4 coords (in $coo)
##   - 4 classifiers (in $fac): 
## # A tibble: 210 x 4
##   var   domes view  ind  
##   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
## 1 Aglan cult  VD    O10  
## 2 Aglan cult  VL    O10  
## 3 Aglan cult  VD    O11  
## 4 Aglan cult  VL    O11  
## 5 Aglan cult  VD    O12  
## 6 Aglan cult  VL    O12  
## # ... with 204 more rows
##   - also: $ldk</code></pre>
<p>Now, we gonna calculate <code>opoly</code> on it and plot the result of the PCA. Notice how consistent is the grammar and the objects obtained:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">op &lt;-<span class="st"> </span><span class="kw"><a href="../reference/opoly.html">opoly</a></span>(olea)           <span class="co"># orthogonal polynomials</span>
<span class="kw">class</span>(op)                   <span class="co"># an OpnCoe, but also a Coe</span>
op.p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>(op)             <span class="co"># we calculate a PCA on it</span>
<span class="kw">class</span>(op.p)                 <span class="co"># a PCA object</span>
<span class="kw">plot</span>(<span class="kw"><a href="../reference/PCA.html">PCA</a></span>(op), <span class="op">~</span>domes<span class="op">+</span>var)   <span class="co"># notice the formula interface to combine factors</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<pre><code>## [1] "OpnCoe" "Coe"   
## [1] "PCA"    "prcomp"</code></pre>
<p>But this is perfectly wrong! We merged the two views are if they were different individuals. Momocs can first <code>chop</code> or <code>filter</code> the whole dataset to separate the two views, do morphometrics on them, and <code>combine</code> them afterwards.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">with</span>(olea<span class="op">$</span>fac, <span class="kw">table</span>(view, var))
<span class="co"># we drop 'Cypre' since there is no VL for 'Cypre' var</span>
olea <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filter.html">filter</a></span>(var <span class="op">!=</span><span class="st"> "Cypre"</span>) <span class="op">%&gt;%</span><span class="st">              </span>
<span class="st">  </span><span class="co"># split, do morphometrics, combine</span>
<span class="st">  </span><span class="kw"><a href="../reference/chop.html">chop</a></span>(<span class="op">~</span>view) <span class="op">%&gt;%</span><span class="st"> </span>opoly <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/combine.html">combine</a></span>() <span class="op">%T&gt;%</span>
<span class="st">   </span><span class="co"># we print the OpnCoe object, then resume to the pipe</span>
<span class="st">  </span><span class="kw">print</span>() <span class="op">%&gt;%</span>
<span class="st">  </span><span class="co"># note the two views in the morphospace</span>
<span class="st">  </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/plot_PCA.html">plot_PCA</a></span>(<span class="op">~</span>var)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<pre><code>##     var
## view Aglan Cypre MouBo1 PicMa
##   VD    30    30     30    30
##   VL    30     0     30    30
## An OpnCoe object [ combined: opoly + opoly analyses ]
## --------------------
##  - $coe: 90 open outlines described
##  - $baseline1: (-0.5; 0; -0.5; 0)
##  - $baseline2: (0.5; 0; 0.5; 0)
## # A tibble: 90 x 4
##   var   domes view  ind  
##   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
## 1 Aglan cult  VD    O10  
## 2 Aglan cult  VD    O11  
## 3 Aglan cult  VD    O12  
## 4 Aglan cult  VD    O13  
## 5 Aglan cult  VD    O14  
## 6 Aglan cult  VD    O15  
## # ... with 84 more rows</code></pre>
<p>Now the PCA is done on the combination of two <code>OpnCoe</code> objects, each one resulting from an independant <code>opoly</code> call. That is the meaning of the <code>[ combined: opoly + opoly analyses ]</code> printed by the pipe above. Momocs can combine up to four different views.</p>
</div>
<div id="configuration-of-landmarks" class="section level3">
<h3 class="hasAnchor">
<a href="#configuration-of-landmarks" class="anchor"></a>Configuration of landmarks</h3>
<p>_Landmarks methods are still quite experimental (i.e. not tested extensively)</p>
<p>Let’s have a look to graphics facilities and apply a full generalized Procrustes adjustment:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/pile.html">pile</a></span>(wings)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">w.al &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fgProcrustes.html">fgProcrustes</a></span>(wings)
<span class="kw"><a href="../reference/pile.html">pile</a></span>(w.al)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-22-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># PCA</span>
<span class="kw"><a href="../reference/PCA.html">PCA</a></span>(w.al) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="dv">1</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-22-3.png" width="672"></p>
<p>Sliding landmarks are supported and rely on <code>geomorph</code> package by D.C. Adams and E. Otarola-Castillo.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/pile.html">pile</a></span>(chaff)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-23-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chaff.al &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fgsProcrustes.html">fgsProcrustes</a></span>(chaff)
<span class="kw"><a href="../reference/pile.html">pile</a></span>(chaff.al)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-23-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">chaff.al <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="op">~</span>taxa, <span class="dt">chull.filled=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-23-3.png" width="672"></p>
<pre><code>## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |=============                                                    |  20%
  |                                                                       
  |==========================                                       |  40%
  |                                                                       
  |=======================================                          |  60%
  |                                                                       
  |====================================================             |  80%
  |                                                                       
  |=================================================================| 100%</code></pre>
<p>Again, the grammar is consistent for landmarks.</p>
</div>
<div id="traditional-morphometrics" class="section level3">
<h3 class="hasAnchor">
<a href="#traditional-morphometrics" class="anchor"></a>Traditional morphometrics</h3>
<p>Traditional morphometrics lose geometries: from the variables, you can’t unambiguously reconstruct the shape. Every shape is described by a combination of measurements, (inter landmark distance, quantitative variables, scalar descriptor, etc.)</p>
<p>Momocs provides some basics utilities to work with such objects in the <code>TraCoe</code> class. There is not <code>TraCoo</code> per se but it can be obtained from any <code>Coo</code> with the <code>measure</code> method. Let’s take the <code>hearts</code> dataset that comes from handdrawn heart shapes from my former colleagues at the French Intitute of Pondicherry:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hearts
<span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(hearts, <span class="dt">fac=</span><span class="st">"aut"</span>, <span class="dt">names=</span><span class="st">"aut"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-24-1.png" width="672"></p>
<pre><code>## Out (outlines)
##   - 240 outlines, 80 +/- 0 coords (in $coo)
##   - 1 classifiers (in $fac): 
## # A tibble: 240 x 1
##   aut  
##   &lt;fct&gt;
## 1 ced  
## 2 ced  
## 3 ced  
## 4 ced  
## 5 ced  
## 6 ced  
## # ... with 234 more rows
##   - also: $ldk</code></pre>
<p>Notice that there are 4 landmarks defined on them. Such landmarks on outlines can be: defined with<code><a href="../reference/def_ldk.html">def_ldk()</a></code>, retrieved with <code><a href="../reference/get_ldk.html">get_ldk()</a></code>, and overall used to align outlines with <code><a href="../reference/fgProcrustes.html">fgProcrustes()</a></code>. You can compare: <code>hearts %&gt;% stack()</code> with <code>hearts %&gt;% fgProcrustes() %&gt;% coo_slide(ldk=1) %&gt;% stack()</code>.</p>
<p>Let’s describe these hearts with scalar descriptors: area, circularity and the distance between the 1st and the 3rd bumps of the hearts. <code>measure</code> is of great help. Note the loadings.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ht &lt;-<span class="st"> </span><span class="kw"><a href="../reference/measure.html">measure</a></span>(hearts, coo_area, coo_circularity, <span class="kw"><a href="../reference/d.html">d</a></span>(<span class="dv">1</span>, <span class="dv">3</span>))
<span class="kw">class</span>(ht)
ht<span class="op">$</span>coe
ht <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"aut"</span>, <span class="dt">pch=</span><span class="dv">20</span>, <span class="dt">ellipsesax=</span>F, <span class="dt">ellipse=</span>T, <span class="dt">loadings=</span>T)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-25-1.png" width="672"></p>
<pre><code>## [1] "TraCoe" "Coe"   
## # A tibble: 240 x 3
##     area circularity   d1.3
##  * &lt;dbl&gt;       &lt;dbl&gt;  &lt;dbl&gt;
##  1 0.627        14.0 0.0695
##  2 0.593        14.5 0.0639
##  3 0.611        16.1 0.0699
##  4 0.539        15.7 0.0627
##  5 0.541        13.6 0.0613
##  6 0.611        16.0 0.0692
##  7 0.617        16.5 0.0755
##  8 0.535        15.8 0.0655
##  9 0.521        14.3 0.0622
## 10 0.546        15.2 0.0642
## # ... with 230 more rows</code></pre>
<p>Again, there are plenty of scalar descriptors of shape, which names starts with <code>coo_*</code>, <code>apropos("coo_")</code>.</p>
<p>Such a <code>TraCoe</code> is provided in the <code>flower</code> dataset which is simply a rearranged <code>iris</code>. Once again, note the grammar consistency.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">flower
flower <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"sp"</span>, <span class="dt">loadings=</span><span class="ot">TRUE</span>, <span class="dt">contour=</span><span class="ot">TRUE</span>, <span class="dt">lev.contour=</span><span class="dv">5</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-26-1.png" width="672"></p>
<pre><code>## A TraCoe object --------------------
##  - $coe: 150 shapes described with 4 variables
## # A tibble: 150 x 1
##   sp    
##   &lt;fct&gt; 
## 1 setosa
## 2 setosa
## 3 setosa
## 4 setosa
## 5 setosa
## 6 setosa
## # ... with 144 more rows</code></pre>
<p>Note that, by default, PCA on <code>TraCoe</code> object first centers and scales variables. This can be changed, see <code><a href="../reference/PCA.html">?PCA</a></code>.</p>
</div>
</div>
<div id="multivariate-statistics" class="section level2">
<h2 class="hasAnchor">
<a href="#multivariate-statistics" class="anchor"></a>Multivariate statistics</h2>
<p>This section will mainly be illustrated with <code>bot</code>, and consequently outline analysis, but it works the same on any <code>Coe</code> object, resulting from open outlines, configuration of landmarks, traditional morphometrics, etc.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.f &lt;-<span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(bot)</code></pre></div>
<div id="pca-principal-component-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#pca-principal-component-analysis" class="anchor"></a><code>PCA</code>: Principal Component Analysis</h3>
<p>Let’s see the main components of shape variability with a Principal Component Analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.p &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>(bot.f)
<span class="kw">plot</span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-28-1.png" width="672"></p>
<p>Morphological spaces are reconstructed on the fly with <code>plot.PCA</code>. We call it <code>plot.PCA</code> because it uses the familiar <code>plot</code> but on the particular <code>PCA</code> class (type <code>class(bot.p)</code>). We may want to display the two groups saved in <code>bot$fac</code>. Just type the id of the column or its name.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="st">"type"</span>) <span class="co"># equivalent to plot(bot.p, 1)</span></code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-29-1.png" width="672"></p>
<p><code>plot.PCA</code> has many arguments and examples below will give a glimpse, but see <code><a href="../reference/plot.PCA.html">?plot.PCA</a></code> for the full list.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="dv">1</span>, <span class="dt">ellipses=</span><span class="ot">TRUE</span>, <span class="dt">ellipsesax =</span> <span class="ot">FALSE</span>, <span class="dt">pch=</span><span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">5</span>))</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-30-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="dv">1</span>, <span class="dt">chull=</span><span class="ot">TRUE</span>, <span class="dt">pos.shp =</span> <span class="st">"full_axes"</span>, <span class="dt">abbreviate.labelsgroups =</span> <span class="ot">TRUE</span>, <span class="dt">points=</span><span class="ot">FALSE</span>, <span class="dt">labelspoints =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-30-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(bot.p, <span class="dv">1</span>, <span class="dt">pos.shp=</span><span class="st">"circle"</span>, <span class="dt">stars=</span><span class="ot">TRUE</span>, <span class="dt">chull.filled=</span><span class="ot">TRUE</span>, <span class="dt">palette=</span>col_spring)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-30-3.png" width="672"></p>
<p>You have a <code>ggplot2</code> alternative, that is not completely consistent:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot2(bot.p, "type") # deprecated for the moment</span></code></pre></div>
<p>And other helper functions for the <code>PCA</code> class:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/scree.html">scree</a></span>(bot.p)
<span class="kw"><a href="../reference/scree.html">scree_plot</a></span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-32-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">boxplot</span>(bot.p, <span class="dv">1</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-32-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/PCcontrib.html">PCcontrib</a></span>(bot.p)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-32-3.png" width="672"></p>
<pre><code>## # A tibble: 10 x 3
##    axis  proportion cumsum
##    &lt;ord&gt;      &lt;dbl&gt;  &lt;dbl&gt;
##  1 1        0.761    0.761
##  2 2        0.170    0.931
##  3 3        0.0294   0.960
##  4 4        0.0135   0.974
##  5 5        0.00860  0.982
##  6 6        0.00719  0.990
##  7 7        0.00306  0.993
##  8 8        0.00190  0.994
##  9 9        0.00159  0.996
## 10 10       0.00122  0.997</code></pre>
<p>By the way, you can use Momocs plotters to plot non-morphometric datasets. Using a <code>TraCoe</code> object is an option, but <code>PCA</code> also works fine. Let’s see an example with <code>iris</code> dataset:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/TraCoe.html">TraCoe</a></span>(iris[, <span class="op">-</span><span class="dv">5</span>], <span class="dt">fac=</span><span class="kw">data.frame</span>(<span class="dt">sp=</span>iris<span class="op">$</span>Species)) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"sp"</span>, <span class="dt">loadings=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-33-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># or</span>
<span class="kw"><a href="../reference/PCA.html">PCA</a></span>(iris[, <span class="op">-</span><span class="dv">5</span>], <span class="dt">fac=</span><span class="kw">data.frame</span>(<span class="dt">sp=</span>iris<span class="op">$</span>Species)) <span class="op">%&gt;%</span>
<span class="st">  </span><span class="kw">plot</span>(<span class="st">"sp"</span>, <span class="dt">chull=</span><span class="ot">TRUE</span>, <span class="dt">ellipses=</span><span class="ot">TRUE</span>, <span class="dt">conf_ellipses =</span> <span class="fl">0.9</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-33-2.png" width="672"></p>
</div>
<div id="lda-linear-discriminant-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#lda-linear-discriminant-analysis" class="anchor"></a><code>LDA</code>: Linear Discriminant Analysis</h3>
<p>We can also calculate a Linear Discriminant Analysis on the PCA scores, or on the <code>Coe</code> object, directly on the matrix of coefficients (and results may be better yet we may encounter collinearity between variables). Try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#LDA(bot.f, 1)</span>
<span class="co"># we work on PCA scores</span>
bot.l &lt;-<span class="st"> </span><span class="kw"><a href="../reference/LDA.html">LDA</a></span>(bot.p, <span class="dv">1</span>)
<span class="co"># print a summary, along with the leave-one-out cross-validation table. </span>
bot.l 
<span class="co"># a much more detailed summary</span>
bot.l <span class="op">%&gt;%</span><span class="st"> </span>summary
<span class="co"># plot.LDA works pretty much with the same grammar as plot.PCA</span>
<span class="co"># here we only have one LD</span>
<span class="kw">plot</span>(bot.l)
<span class="co"># plot the cross-validation table</span>
<span class="kw"><a href="../reference/plot_CV.html">plot_CV</a></span>(bot.l)  <span class="co"># tabular version</span>
<span class="kw"><a href="../reference/plot_CV.html">plot_CV</a></span>(bot.l, <span class="dt">freq=</span><span class="ot">TRUE</span>) <span class="co"># frequency table</span>
<span class="kw"><a href="../reference/plot_CV2.html">plot_CV2</a></span>(bot.l) <span class="co"># arrays version</span></code></pre></div>
<p>Finally, if you need to classify using LDA methods, see <code>classify</code>.</p>
</div>
<div id="manova-multivariate-analysis-of-covariace" class="section level3">
<h3 class="hasAnchor">
<a href="#manova-multivariate-analysis-of-covariace" class="anchor"></a><code>MANOVA</code>: Multivariate Analysis of (co)variace</h3>
<p>We can test for a difference in the distribution of PC scores with:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/MANOVA.html">MANOVA</a></span>(bot.p, <span class="st">"type"</span>)</code></pre></div>
<pre><code>##           Df Hotelling-Lawley approx F num Df den Df   Pr(&gt;F)    
## fac        1           2.8388   12.977      7     32 8.88e-08 ***
## Residuals 38                                                     
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
<p>We can also calculate pairwise combination between every levels of a <code>fac</code>. Here we just have two levels, so a single pairwise combination but the syntax is:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/MANOVA_PW.html">MANOVA_PW</a></span>(bot.p, <span class="st">"type"</span>)</code></pre></div>
<pre><code>## $stars.tab
##      beer whisky
## beer      ***   
## 
## $summary (see also $manovas)
##               Df Pillai approx F num Df den Df   Pr(&gt;F)
## beer - whisky  1 0.7395    12.98      7     32 8.88e-08</code></pre>
<p>If we want a MANCOVA instead :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mutate.html">mutate</a></span>(<span class="dt">cs=</span><span class="kw"><a href="../reference/coo_centsize.html">coo_centsize</a></span>(.))
bot <span class="op">%&gt;%</span><span class="st"> </span>efourier <span class="op">%&gt;%</span><span class="st"> </span>PCA <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/MANOVA.html">MANOVA</a></span>(<span class="st">"cs"</span>)</code></pre></div>
<pre><code>##           Df Hotelling-Lawley approx F num Df den Df Pr(&gt;F)
## fac        1          0.32229   1.4733      7     32  0.212
## Residuals 38</code></pre>
</div>
<div id="clust-hierarchical-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#clust-hierarchical-clustering" class="anchor"></a><code>CLUST</code>: Hierarchical clustering</h3>
<p>A hierarchical classification now. It relies on <code>dist</code> + <code>hclust</code> + <code><a href="http://www.rdocumentation.org/packages/ape/topics/plot.phylo">ape::plot.phylo</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/CLUST.html">CLUST</a></span>(bot.p, <span class="dv">1</span>)</code></pre></div>
<p>Monophyly is plotted by default. Many options can be found in <code><a href="../reference/CLUST.html">?CLUST</a></code></p>
</div>
<div id="kmeans-k-means-clustering" class="section level3">
<h3 class="hasAnchor">
<a href="#kmeans-k-means-clustering" class="anchor"></a><code>KMEANS</code>: K-means clustering</h3>
<p>A very minimal k-means clustering is implemented:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/KMEANS.html">KMEANS</a></span>(bot.p, <span class="dt">centers =</span> <span class="dv">5</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-39-1.png" width="672"></p>
<pre><code>## K-means clustering with 5 clusters of sizes 6, 18, 6, 5, 5
## 
## Cluster means:
##           PC1          PC2
## 1  0.10184224 -0.028875902
## 2 -0.04433327  0.008000365
## 3 -0.04573875 -0.031541607
## 4  0.02497106 -0.006671549
## 5  0.06730453  0.050371247
## 
## Clustering vector:
##         brahma          caney         chimay         corona    deusventrue 
##              3              2              1              2              2 
##          duvel   franziskaner     grimbergen        guiness     hoegardeen 
##              1              3              4              4              2 
##        jupiler     kingfisher       latrappe lindemanskriek    nicechouffe 
##              2              2              1              2              2 
##     pecheresse   sierranevada     tanglefoot          tauro      westmalle 
##              2              4              1              2              2 
##          amrut    ballantines      bushmills         chivas        dalmore 
##              2              1              3              5              5 
##   famousgrouse    glendronach   glenmorangie   highlandpark    jackdaniels 
##              3              2              2              1              4 
##             jb  johnniewalker       magallan     makersmark           oban 
##              2              3              3              5              2 
##     oldpotrero      redbreast         tamdhu     wildturkey         yoichi 
##              5              5              2              2              4 
## 
## Within cluster sum of squares by cluster:
## [1] 0.006121728 0.006537571 0.001860776 0.001916822 0.002720829
##  (between_SS / total_SS =  89.4 %)
## 
## Available components:
## 
## [1] "cluster"      "centers"      "totss"        "withinss"    
## [5] "tot.withinss" "betweenss"    "size"         "iter"        
## [9] "ifault"</code></pre>
</div>
<div id="mshapes-mean-shapes" class="section level3">
<h3 class="hasAnchor">
<a href="#mshapes-mean-shapes" class="anchor"></a><code>mshapes</code>: Mean shapes</h3>
<p>We can retrieve the mean shapes, group wise (if a <code>fac</code> is specified), or the global mean shape (if omitted). It works from the <code>Coe</code> object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mean shape</span>
bot.f <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mshapes.html">mshapes</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-40-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mean shape, per group</span>
bot.ms &lt;-<span class="st"> </span><span class="kw"><a href="../reference/mshapes.html">mshapes</a></span>(bot.f, <span class="dv">1</span>)
beer   &lt;-<span class="st"> </span>bot.ms<span class="op">$</span>shp<span class="op">$</span>beer    <span class="op">%T&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(<span class="dt">border=</span><span class="st">"blue"</span>)
whisky &lt;-<span class="st"> </span>bot.ms<span class="op">$</span>shp<span class="op">$</span>whisky  <span class="op">%T&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/coo_draw.html">coo_draw</a></span>(<span class="dt">border=</span><span class="st">"red"</span>)
<span class="kw">legend</span>(<span class="st">"topright"</span>, <span class="dt">lwd=</span><span class="dv">1</span>,
       <span class="dt">col=</span><span class="kw">c</span>(<span class="st">"blue"</span>, <span class="st">"red"</span>), <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">"beer"</span>, <span class="st">"whisky"</span>))</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-40-2.png" width="672"></p>
<p>We can also plot a pairwise comparison of them:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># leaves &lt;- shapes %&gt;% slice(grep("leaf", names(shapes))) %$% coo</span>
<span class="co"># leaves %&gt;% plot_mshapes(col2="#0000FF")</span>

<span class="co"># or from mshapes directly</span>
bot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dv">6</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mshapes.html">mshapes</a></span>(<span class="st">"type"</span>) <span class="op">%&gt;%</span><span class="st"> </span>plot_mshapes</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-41-1.png" width="672"></p>
</div>
</div>
<div id="manipulating-objects" class="section level2">
<h2 class="hasAnchor">
<a href="#manipulating-objects" class="anchor"></a>Manipulating objects</h2>
<div id="dplyr-verbs" class="section level3">
<h3 class="hasAnchor">
<a href="#dplyr-verbs" class="anchor"></a><code>dplyr</code> verbs</h3>
<p>One common yet boring task of morphometrics consists in handling datasets: add new information, remove some individuals, etc.</p>
<p>Momocs adapts <a href="https://CRAN.R-project.org/package=dplyr">dplyr</a> verbs to its objects, and add new ones. If you have never heard of dplyr, let’s have a look to its introduction <a href="https://CRAN.R-project.org/package=dplyr">there</a>, this may change your (R) life.</p>
<p>Basics verbs are implemented; try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(olea)
olea</code></pre></div>
<pre><code>## Opn (curves)
##   - 210 curves, 99 +/- 4 coords (in $coo)
##   - 4 classifiers (in $fac): 
## # A tibble: 210 x 4
##   var   domes view  ind  
##   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
## 1 Aglan cult  VD    O10  
## 2 Aglan cult  VL    O10  
## 3 Aglan cult  VD    O11  
## 4 Aglan cult  VL    O11  
## 5 Aglan cult  VD    O12  
## 6 Aglan cult  VL    O12  
## # ... with 204 more rows
##   - also: $ldk</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># slice: select individuals based on their position</span>
<span class="kw"><a href="../reference/slice.html">slice</a></span>(olea, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>)</code></pre></div>
<pre><code>## Opn (curves)
##   - 5 curves, 99 +/- 4 coords (in $coo)
##   - 4 classifiers (in $fac): 
## # A tibble: 5 x 4
##   var   domes view  ind  
##   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
## 1 Aglan cult  VD    O10  
## 2 Aglan cult  VL    O10  
## 3 Aglan cult  VD    O11  
## 4 Aglan cult  VL    O11  
## 5 Aglan cult  VD    O12  
##   - also: $ldk</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/slice.html">slice</a></span>(olea, <span class="op">-</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>))</code></pre></div>
<pre><code>## Opn (curves)
##   - 110 curves, 99 +/- 4 coords (in $coo)
##   - 4 classifiers (in $fac): 
## # A tibble: 110 x 4
##   var   domes view  ind  
##   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
## 1 PicMa cult  VD    O15  
## 2 PicMa cult  VL    O15  
## 3 PicMa cult  VD    O16  
## 4 PicMa cult  VL    O16  
## 5 PicMa cult  VD    O17  
## 6 PicMa cult  VL    O17  
## # ... with 104 more rows
##   - also: $ldk</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter: select individual based on a logical condition</span>
<span class="kw"><a href="../reference/filter.html">filter</a></span>(olea, domes<span class="op">==</span><span class="st">"cult"</span>, view<span class="op">!=</span><span class="st">"VD"</span>)</code></pre></div>
<pre><code>## Opn (curves)
##   - 60 curves, 97 +/- 4 coords (in $coo)
##   - 4 classifiers (in $fac): 
## # A tibble: 60 x 4
##   var   domes view  ind  
##   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
## 1 Aglan cult  VL    O10  
## 2 Aglan cult  VL    O11  
## 3 Aglan cult  VL    O12  
## 4 Aglan cult  VL    O13  
## 5 Aglan cult  VL    O14  
## 6 Aglan cult  VL    O15  
## # ... with 54 more rows
##   - also: $ldk</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select: pick, reorder columns from the $fac</span>
<span class="kw"><a href="../reference/select.html">select</a></span>(olea, <span class="dv">1</span>, <span class="dt">Ind=</span>ind)</code></pre></div>
<pre><code>## Opn (curves)
##   - 210 curves, 99 +/- 4 coords (in $coo)
##   - 2 classifiers (in $fac): 
## # A tibble: 210 x 2
##   var   Ind  
##   &lt;fct&gt; &lt;fct&gt;
## 1 Aglan O10  
## 2 Aglan O10  
## 3 Aglan O11  
## 4 Aglan O11  
## 5 Aglan O12  
## 6 Aglan O12  
## # ... with 204 more rows
##   - also: $ldk</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># mutate: add new columns</span>
<span class="kw"><a href="../reference/mutate.html">mutate</a></span>(olea, <span class="dt">fake=</span><span class="kw">factor</span>(<span class="kw">rep</span>(letters[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], <span class="dt">each=</span><span class="dv">105</span>)))</code></pre></div>
<pre><code>## Opn (curves)
##   - 210 curves, 99 +/- 4 coords (in $coo)
##   - 5 classifiers (in $fac): 
## # A tibble: 210 x 5
##   var   domes view  ind   fake 
##   &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt; &lt;fct&gt;
## 1 Aglan cult  VD    O10   a    
## 2 Aglan cult  VL    O10   a    
## 3 Aglan cult  VD    O11   a    
## 4 Aglan cult  VL    O11   a    
## 5 Aglan cult  VD    O12   a    
## 6 Aglan cult  VL    O12   a    
## # ... with 204 more rows
##   - also: $ldk</code></pre>
<p>And you can pipe those operations: say, we only want dorsal views from domesticated individuals, for a (renamed) ‘status’ column, and drop the ‘ind’ column:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">olea <span class="op">%&gt;%</span><span class="st"> </span>
<span class="kw"><a href="../reference/filter.html">filter</a></span>(domes<span class="op">==</span><span class="st">"cult"</span>, view<span class="op">==</span><span class="st">"VD"</span>) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="kw"><a href="../reference/rename.html">rename</a></span>(<span class="dt">domesticated=</span>domes) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="kw"><a href="../reference/select.html">select</a></span>(<span class="op">-</span>ind)</code></pre></div>
<pre><code>## Opn (curves)
##   - 90 curves, 100 +/- 2 coords (in $coo)
##   - 3 classifiers (in $fac): 
## # A tibble: 90 x 3
##   var   domesticated view 
##   &lt;fct&gt; &lt;fct&gt;        &lt;fct&gt;
## 1 Aglan cult         VD   
## 2 Aglan cult         VD   
## 3 Aglan cult         VD   
## 4 Aglan cult         VD   
## 5 Aglan cult         VD   
## 6 Aglan cult         VD   
## # ... with 84 more rows
##   - also: $ldk</code></pre>
<p>This should save some headaches.</p>
</div>
<div id="new-verbs-ala-dplyr" class="section level3">
<h3 class="hasAnchor">
<a href="#new-verbs-ala-dplyr" class="anchor"></a>New verbs ala <code>dplyr</code>
</h3>
<p>New verbs are implemented: for instance, you can <code>chop</code> (a rougher slicing) according to a condition: this will create a list, on which you can apply further operations, then <code>combine</code> it back. This is particularly useful when you want to apply independant treatments to different partitions, eg orthogonal views of your model. Prior to this, we can use <code>table</code> to cross-tabulate data from <code>$fac</code>. We could have done the first step of what follows with <code>rm_uncomplete</code> that drops (if any) missing data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">with</span>(olea<span class="op">$</span>fac, <span class="kw">table</span>(var, view))
<span class="co"># we drop 'Cypre' since there is no VL for 'Cypre' var</span>
olea <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filter.html">filter</a></span>(var <span class="op">!=</span><span class="st"> "Cypre"</span>) <span class="op">%&gt;%</span><span class="st">              </span>
<span class="st">  </span><span class="co"># split, do morphometrics, combine</span>
<span class="st">  </span><span class="kw"><a href="../reference/chop.html">chop</a></span>(<span class="op">~</span>view) <span class="op">%&gt;%</span><span class="st"> </span>opoly <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/combine.html">combine</a></span>() <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># note the two views in the morphospace</span>
<span class="st">  </span><span class="kw"><a href="../reference/PCA.html">PCA</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">plot</span>(<span class="st">"var"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-44-1.png" width="672"></p>
<pre><code>##         view
## var      VD VL
##   Aglan  30 30
##   Cypre  30  0
##   MouBo1 30 30
##   PicMa  30 30</code></pre>
<p>Also if you need some <code>group_by</code> operation, you can use the <code>fac</code> <code>data.frame</code>. For instance if you want to <code>arrange</code> based on a column and retain the top 10 shapes, then you can:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">retain &lt;-<span class="st"> </span>x<span class="op">$</span>fac <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span><span class="co"># here we sort and create an id</span>
<span class="st">  </span><span class="kw"><a href="../reference/arrange.html">arrange</a></span>(your_column) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/mutate.html">mutate</a></span>(<span class="dt">.id=</span><span class="dv">1</span><span class="op">:</span><span class="kw">n</span>()) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(taxa) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/slice.html">slice</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>) <span class="op">%$%</span><span class="st"> </span>.id
<span class="co"># it worked!</span>
x <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/slice.html">slice</a></span>(retain) <span class="op">%$%</span><span class="st"> </span><span class="kw">table</span>(fac<span class="op">$</span>taxa)</code></pre></div>
</div>
<div id="various-helpers" class="section level3">
<h3 class="hasAnchor">
<a href="#various-helpers" class="anchor"></a>Various helpers</h3>
<p>Some methods help, on <code>Coe</code> objects to: * select groups with at least a certain number of individuals in them: <code>at_least</code> * removes outliers : <code>which_out</code> * sample a given number: <code>sample_n</code>; * sample a given proportion: <code>sample_frac</code>; * generate new individuals based on calibrated Gaussian coefficient generation: <code>breed</code>; * generate new individuals based on permutations: <code>perm</code>.</p>
<p>Several shortcuts are implemented on <code>Coo</code> and <code>Coe</code> objects: * <code>names</code> returns shape names; * <code>length</code> returns their number; * <code>table</code> cross-tabulates the<code>$fac</code> component; * <code>Ntable</code> does the same job and plots a confusion matrix; * <code>[]</code> extracts one (or more) shape; * <code>$</code> can access either a shape name or a column name for the <code>$fac</code>.</p>
<p>Try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(bot)
<span class="kw">length</span>(bot)
<span class="kw">table</span>(olea, <span class="st">"var"</span>, <span class="st">"domes"</span>)
bot[<span class="dv">1</span>]
bot[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]
bot<span class="op">$</span>brahma
bot<span class="op">$</span>type</code></pre></div>
</div>
</div>
<div id="babel-importexport" class="section level2">
<h2 class="hasAnchor">
<a href="#babel-importexport" class="anchor"></a>Babel import/export</h2>
<p>There are various morphometrics formats in the wild, almost as much as softwares. Momocs tries to create bridges between them, all gathered in the <strong>Babel</strong> family.</p>
<div id="bridges-within-r" class="section level3">
<h3 class="hasAnchor">
<a href="#bridges-within-r" class="anchor"></a>Bridges within R</h3>
<p>You can convert from/to array, matrix, list or data.frame with the functions <code>{a, m, l, d}2{a, m, l, d}</code>. For instance, <code>l2a</code> converts a list into an array that you can use with <code>geomorph</code>; <code>a2l</code> does the inverse operation.</p>
<p>Imagine you want to import pupfish from <code>geomorph</code> as a <code>Ldk</code> object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(geomorph)
<span class="kw">data</span>(pupfish)
<span class="kw">str</span>(pupfish)
<span class="co"># so $coords will become $coo, and</span>
<span class="co"># all other components will be turned into a data.frame to feed $fac</span>
<span class="co"># with a single line</span>
<span class="kw"><a href="../reference/Ldk.html">Ldk</a></span>(<span class="dt">coo=</span>pupfish<span class="op">$</span>coords <span class="op">%&gt;%</span><span class="st"> </span>a2l,
<span class="dt">fac=</span>pupfish[<span class="op">-</span><span class="dv">1</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>())</code></pre></div>
</div>
<div id="import-from-stereomorph" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-stereomorph" class="anchor"></a>Import from StereoMorph</h3>
<p>If you use this excellent package by Olsen to digitize landmarks and curves, you can import them, from the files produced with the functions <code>import_StereoMorph_ldk</code> and <code>import_StereoMorph_curve</code>.</p>
</div>
<div id="import-from-tps-and-other-digitizing-softwares" class="section level3">
<h3 class="hasAnchor">
<a href="#import-from-tps-and-other-digitizing-softwares" class="anchor"></a>Import from tps and other digitizing softwares</h3>
<ul>
<li>
<code>.tps</code> files can be read with <code>import_tps</code>
</li>
<li>
<code>.nts</code> files an be read with <code>nts2Coo</code> (will be turned into <code>import_nts</code> soon)</li>
</ul>
</div>
<div id="direct-build-of-coe-objects-youre-not-bound-with-momocs-from-the-shapes-step-ie-you-do-not-have-to-start-from-coo-objects--for-instance-if-you-have-a-matrix-of-coefficients-you-can-directly-build-an-outcoe-with-the-builder-see-below--same-approach-for-opncoe-and-tracoe-have-a-look-to-the-help-files-of-these-builders-" class="section level3">
<h3 class="hasAnchor">
<a href="#direct-build-of-coe-objects-youre-not-bound-with-momocs-from-the-shapes-step-ie-you-do-not-have-to-start-from-coo-objects--for-instance-if-you-have-a-matrix-of-coefficients-you-can-directly-build-an-outcoe-with-the-builder-see-below--same-approach-for-opncoe-and-tracoe-have-a-look-to-the-help-files-of-these-builders-" class="anchor"></a>Direct build of <em>Coe objects </em> You’re not bound with Momocs from the “shapes” step, ie you do not have to start from <code>Coo</code> objects. For instance if you have a matrix of coefficients, you can directly build an <code>OutCoe</code> with the builder (see below). Same approach for <code>OpnCoe</code> and <code>TraCoe</code>; have a look to the help files of these builders.</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we simulate an imported matrix of coordinates, eg from a .csv</span>
coeffs_from_the_wild &lt;-<span class="st"> </span>bot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/efourier.html">efourier</a></span>(<span class="dv">6</span>) <span class="op">%$%</span><span class="st"> </span>coe
coeffs_in_Momocs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/OutCoe.html">OutCoe</a></span>(<span class="dt">coe=</span>coeffs_from_the_wild, <span class="dt">method=</span><span class="st">"efourier"</span>, <span class="dt">norm=</span><span class="ot">TRUE</span>)
coeffs_in_Momocs <span class="op">%&gt;%</span><span class="st"> </span>PCA <span class="op">%&gt;%</span><span class="st"> </span>plot</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-48-1.png" width="672"></p>
</div>
<div id="import-misc" class="section level3">
<h3 class="hasAnchor">
<a href="#import-misc" class="anchor"></a>Import misc</h3>
<ul>
<li>If covariables are encoded in filenames, which is a good practice, eg if you have files named <code>spA_group7_ind4_VL.{txt|jpg|etc.}</code>, use <code>lf_structure</code>;</li>
<li>If you need rewriting rules on your <code>$fac</code>, <code>rw_rule</code> is your friend;</li>
<li>If you need to rescale imported coordinates, see <code>rescale</code>;</li>
<li>If you need to tie images (eg outlines) and <code>.txt</code> (eg coordinates of landmarks on them), see <code>tie_jpg_txt</code>.</li>
</ul>
</div>
<div id="save-from-r" class="section level3">
<h3 class="hasAnchor">
<a href="#save-from-r" class="anchor"></a>Save from R</h3>
<p>The best way to save a Momocs object is probably to use the base <code>save</code> function. You can call it back afterwards with <code>load</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">save</span>(bot, <span class="dt">file=</span><span class="st">"Bottles.rda"</span>)
<span class="co"># closing R, going to the beach, back at work</span>
<span class="kw">load</span>(<span class="st">"Bottles.rda"</span>) <span class="co"># bot is back</span></code></pre></div>
</div>
<div id="export-from-r" class="section level3">
<h3 class="hasAnchor">
<a href="#export-from-r" class="anchor"></a>Export from R</h3>
<p>Any Momocs object, <code>Coo</code>s, <code>Coe</code>s, <code>PCA</code>s, etc. can be turned into a <code>data.frame</code> with <code>as_df</code>. Work with <code>dplyr</code>, <code>ggplot2</code> is made easy and you can export it as <code>.txt</code>, <code>.csv</code> “by hand” or use the <code>export</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot <span class="op">%&gt;%</span><span class="st"> </span>as_df <span class="co"># then %&gt;% write.table</span>
bot <span class="op">%&gt;%</span><span class="st"> </span>efourier <span class="op">%&gt;%</span><span class="st"> </span>export
bot <span class="op">%&gt;%</span><span class="st"> </span>efourier <span class="op">%&gt;%</span><span class="st"> </span>PCA <span class="op">%&gt;%</span><span class="st"> </span>export</code></pre></div>
<p>But, of course, you can directly access information within the Momocs objects; try the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># from Coo objects</span>
bot<span class="op">$</span>coo <span class="co"># list of matrices (of xy coordinates)</span>
bot<span class="op">$</span>fac <span class="co"># data.frame for covariates</span>

<span class="co"># from Coe objects</span>
bot.f<span class="op">$</span>coe <span class="co"># matrix of coefficients</span>
bot.f<span class="op">$</span>fac <span class="co"># data.frame for covariates</span>
<span class="kw"><a href="../reference/as_df.html">as_df</a></span>(bot.f)

<span class="co"># from PCA objects</span>
bot.p<span class="op">$</span>x        <span class="co"># scores</span>
bot.p<span class="op">$</span>rotation <span class="co"># rotation matrix</span></code></pre></div>
</div>
</div>
<div id="graphics" class="section level2">
<h2 class="hasAnchor">
<a href="#graphics" class="anchor"></a>Graphics</h2>
<p>The most common graphics have been already presented in previous sections. But there are many more of them, and they are gathered in the <strong>Graphics</strong> family that can be accessed with … TODO.</p>
<div id="tps_-thin-plate-splines" class="section level3">
<h3 class="hasAnchor">
<a href="#tps_-thin-plate-splines" class="anchor"></a><code>tps_*</code>: Thin Plate Splines</h3>
<p>TPS have not been presented before but here there are:</p>
<pre><code><a href="../reference/tps_grid.html">tps_grid(beer, whisky)
tps_arr(beer, whisky)
tps_iso(beer, whisky)</a></code></pre>
<p>Again, plenty options in <code>?tps_*</code>.</p>
<p>You may also like lolliplots and friends:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_lolli.html">coo_lolli</a></span>(beer, whisky); <span class="kw">title</span>(<span class="st">"coo_lolli"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-52-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/coo_arrows.html">coo_arrows</a></span>(beer, whisky); <span class="kw">title</span>(<span class="st">"coo_arrow"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-52-2.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># an example with coo_ruban</span>
<span class="kw"><a href="../reference/coo_plot.html">coo_plot</a></span>(beer) <span class="co"># to get the first plot </span>
<span class="kw"><a href="../reference/coo_ruban.html">coo_ruban</a></span>(beer, <span class="kw"><a href="../reference/edm.html">edm</a></span>(beer, whisky), <span class="dt">lwd=</span><span class="dv">8</span>) <span class="co"># we add ruban based from deviations</span>
<span class="kw"><a href="../reference/coo_draw.html">coo_draw</a></span>(whisky)
<span class="kw">title</span>(<span class="st">"coo_ruban"</span>)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-52-3.png" width="672"></p>
</div>
<div id="ggplot2" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2" class="anchor"></a><code>ggplot2</code>
</h3>
<p>Many graphical function have their <code>ggplot2</code> counterpart:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/panel.Coo.html">panel</a></span>(bot)</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-53-1.png" width="672"></p>
<p>Otherwise, you can build your own plots. Let’s begin with a simple example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># we build a ggplot object from a shape turned into a data.frame</span>
<span class="kw">library</span>(ggplot2)
shapes[<span class="dv">4</span>] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/bridges.html">m2d</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(x, y) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_path">geom_path</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/coord_fixed">coord_equal</a></span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/labs">labs</a></span>(<span class="dt">title=</span><span class="st">"ggplot2 Meow"</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_minimal</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-54-1.png" width="672"></p>
<p>Here is a more complicated plot that takes profit of <code>as_df</code> methods that turns objects into <code>ggplot2</code>-friendly <code>data.frame</code>s:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bot.p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/as_df.html">as_df</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>() <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x=</span>PC1, <span class="dt">y=</span>PC2, <span class="dt">col=</span>type) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/coord_fixed">coord_equal</a></span>() <span class="op">+</span><span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/geom_density_2d">geom_density2d</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_light</a></span>()</code></pre></div>
<p><img src="Momocs_speed_dating_files/figure-html/unnamed-chunk-55-1.png" width="672"></p>
</div>
<div id="color-palettes" class="section level3">
<h3 class="hasAnchor">
<a href="#color-palettes" class="anchor"></a>Color palettes</h3>
<ul>
<li>Nice palettes are bundled with Momocs, see <code>?palettes</code>.</li>
</ul>
</div>
</div>
<div id="frequently-asked-questions" class="section level2">
<h2 class="hasAnchor">
<a href="#frequently-asked-questions" class="anchor"></a>Frequently asked questions</h2>
<ul>
<li>
<strong>How to access all methods for a given class?</strong>
<ul>
<li>
<code>methods(class="your_class")</code>, eg <code>methods(class="Opn")</code>
</li>
</ul>
</li>
<li>
<strong>How to access methods defined for a particular function?</strong>
<ul>
<li>
<code>methods("your_function")</code>, eg <code>methods("fgProcrustes")</code>
</li>
</ul>
</li>
<li>
<strong>How can I access help for a particular methods?</strong>
<ul>
<li>
<code>function.class</code>, eg <code><a href="../reference/plot.PCA.html">?plot.PCA</a></code>
</li>
</ul>
</li>
<li>
<strong>How can I access the code of a method?</strong>
<ul>
<li>For functions, just type its name, eg <code>lf_structure</code>; for methods add its class, eg <code>plot.PCA</code>
</li>
</ul>
</li>
<li><strong>To be continued</strong></li>
</ul>
<hr>
<ul>
<li>Done with love by Vincent Bonhomme <code>&lt;bonhomme.vincent@gmail.com&gt;</code>.</li>
<li>Help/suggestions/review/critics are welcome.</li>
<li>Sources are <a href="https://github.com/MomX/Momocs/tree/master/vignettes">there</a>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#preliminaries">Preliminaries</a></li>
      <li><a href="#design">Design</a></li>
      <li><a href="#single-shapes">Single shapes</a></li>
      <li><a href="#morphometrics">Morphometrics</a></li>
      <li><a href="#multivariate-statistics">Multivariate statistics</a></li>
      <li><a href="#manipulating-objects">Manipulating objects</a></li>
      <li><a href="#babel-importexport">Babel import/export</a></li>
      <li><a href="#graphics">Graphics</a></li>
      <li><a href="#frequently-asked-questions">Frequently asked questions</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://www.vincentbonhomme.fr">Vincent Bonhomme</a>, Julien Claude.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
